<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Koen Van den Berge and Jeroen Gilis" />


<title>Normalization, feature selection and dimension reduction for the Macosko dataset</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>


<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Single-cell course</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="introSequencingBulk.html">A brief introduction to sequencing technology</a>
    </li>
    <li>
      <a href="singleCellProtocols.html">Two popular single-cell protocols</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Labs
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="lab1_session.html">Lab 1</a>
    </li>
    <li>
      <a href="lab2_session.html">Lab 2</a>
    </li>
    <li>
      <a href="lab3_session.html">Lab 3</a>
    </li>
    <li>
      <a href="lab4_session.html">Lab 4</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/statOmics/Rmd-website">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://statomics.github.io/">statOmics</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
<li role="separator" class="divider"></li>
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Normalization, feature selection and dimension reduction for the Macosko dataset</h1>
<h4 class="author">Koen Van den Berge and Jeroen Gilis</h4>
<h4 class="date">30/11/2021</h4>

</div>


<div id="preamble-installation-of-bioconductor-libraries" class="section level1">
<h1><span class="header-section-number">1</span> Preamble: installation of Bioconductor libraries</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># install BiocManager package if not installed yet.</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="co"># BiocManager is the package installer for Bioconductor software.</span></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</span>
<span id="cb1-4"><a href="#cb1-4"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span></code></pre></div>
<pre><code>## Installing BiocManager [1.30.16] ...
##  OK [linked cache]</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># install packages if not yet installed.</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>pkgs &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;SingleCellExperiment&quot;</span>,</span>
<span id="cb3-3"><a href="#cb3-3"></a>          <span class="st">&quot;ExperimentHub&quot;</span>,</span>
<span id="cb3-4"><a href="#cb3-4"></a>          <span class="st">&quot;edgeR&quot;</span>,</span>
<span id="cb3-5"><a href="#cb3-5"></a>          <span class="st">&quot;biomaRt&quot;</span>,</span>
<span id="cb3-6"><a href="#cb3-6"></a>          <span class="st">&quot;DropletUtils&quot;</span>, </span>
<span id="cb3-7"><a href="#cb3-7"></a>          <span class="st">&quot;scRNAseq&quot;</span>, </span>
<span id="cb3-8"><a href="#cb3-8"></a>          <span class="st">&quot;scater&quot;</span>, </span>
<span id="cb3-9"><a href="#cb3-9"></a>          <span class="st">&quot;scuttle&quot;</span>, </span>
<span id="cb3-10"><a href="#cb3-10"></a>          <span class="st">&quot;scran&quot;</span>,</span>
<span id="cb3-11"><a href="#cb3-11"></a>          <span class="st">&quot;scry&quot;</span>,</span>
<span id="cb3-12"><a href="#cb3-12"></a>          <span class="st">&quot;BiocSingular&quot;</span>, </span>
<span id="cb3-13"><a href="#cb3-13"></a>          <span class="st">&quot;scDblFinder&quot;</span>,</span>
<span id="cb3-14"><a href="#cb3-14"></a>          <span class="st">&quot;glmpca&quot;</span>,</span>
<span id="cb3-15"><a href="#cb3-15"></a>          <span class="st">&quot;PCAtools&quot;</span>,</span>
<span id="cb3-16"><a href="#cb3-16"></a>          <span class="st">&quot;Seurat&quot;</span>)</span>
<span id="cb3-17"><a href="#cb3-17"></a>notInstalled &lt;-<span class="st"> </span>pkgs[<span class="op">!</span>pkgs <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()[,<span class="dv">1</span>]]</span>
<span id="cb3-18"><a href="#cb3-18"></a><span class="cf">if</span>(<span class="kw">length</span>(notInstalled) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>){</span>
<span id="cb3-19"><a href="#cb3-19"></a>  BiocManager<span class="op">::</span><span class="kw">install</span>(notInstalled)</span>
<span id="cb3-20"><a href="#cb3-20"></a>}</span></code></pre></div>
<pre><code>## &#39;getOption(&quot;repos&quot;)&#39; replaces Bioconductor standard repositories, see
## &#39;?repositories&#39; for details
## 
## replacement repositories:
##     CRAN: https://cran.rstudio.com</code></pre>
<pre><code>## Bioconductor version 3.14 (BiocManager 1.30.16), R 4.1.2 (2021-11-01)</code></pre>
<pre><code>## Installing package(s) &#39;BiocVersion&#39;, &#39;SingleCellExperiment&#39;, &#39;ExperimentHub&#39;,
##   &#39;edgeR&#39;, &#39;biomaRt&#39;, &#39;DropletUtils&#39;, &#39;scRNAseq&#39;, &#39;scater&#39;, &#39;scuttle&#39;, &#39;scran&#39;,
##   &#39;scry&#39;, &#39;BiocSingular&#39;, &#39;scDblFinder&#39;, &#39;glmpca&#39;, &#39;PCAtools&#39;, &#39;Seurat&#39;</code></pre>
<pre><code>## also installing the dependencies &#39;bit&#39;, &#39;sys&#39;, &#39;formatR&#39;, &#39;rjson&#39;, &#39;fs&#39;, &#39;GenomeInfoDbData&#39;, &#39;zlibbioc&#39;, &#39;bit64&#39;, &#39;blob&#39;, &#39;memoise&#39;, &#39;plogr&#39;, &#39;DT&#39;, &#39;generics&#39;, &#39;tidyselect&#39;, &#39;assertthat&#39;, &#39;askpass&#39;, &#39;lambda.r&#39;, &#39;futile.options&#39;, &#39;GenomicAlignments&#39;, &#39;restfulr&#39;, &#39;bitops&#39;, &#39;Rhtslib&#39;, &#39;cpp11&#39;, &#39;rprojroot&#39;, &#39;gtools&#39;, &#39;caTools&#39;, &#39;sass&#39;, &#39;jquerylib&#39;, &#39;MatrixGenerics&#39;, &#39;Biobase&#39;, &#39;GenomeInfoDb&#39;, &#39;XVector&#39;, &#39;RSQLite&#39;, &#39;interactiveDisplayBase&#39;, &#39;dplyr&#39;, &#39;dbplyr&#39;, &#39;DBI&#39;, &#39;filelock&#39;, &#39;KEGGREST&#39;, &#39;hms&#39;, &#39;prettyunits&#39;, &#39;openssl&#39;, &#39;futile.logger&#39;, &#39;snow&#39;, &#39;sparseMatrixStats&#39;, &#39;rhdf5filters&#39;, &#39;R.oo&#39;, &#39;R.methodsS3&#39;, &#39;sitmo&#39;, &#39;AnnotationFilter&#39;, &#39;rtracklayer&#39;, &#39;Rsamtools&#39;, &#39;ProtGenerics&#39;, &#39;Biostrings&#39;, &#39;RCurl&#39;, &#39;BiocIO&#39;, &#39;RcppHNSW&#39;, &#39;beeswarm&#39;, &#39;vipor&#39;, &#39;data.table&#39;, &#39;plyr&#39;, &#39;globals&#39;, &#39;listenv&#39;, &#39;parallelly&#39;, &#39;zoo&#39;, &#39;htmlwidgets&#39;, &#39;tidyr&#39;, &#39;lazyeval&#39;, &#39;crosstalk&#39;, &#39;purrr&#39;, &#39;promises&#39;, &#39;here&#39;, &#39;gplots&#39;, &#39;RcppArmadillo&#39;, &#39;httpuv&#39;, &#39;xtable&#39;, &#39;fontawesome&#39;, &#39;sourcetools&#39;, &#39;later&#39;, &#39;commonmark&#39;, &#39;bslib&#39;, &#39;cachem&#39;, &#39;spatstat.data&#39;, &#39;spatstat.utils&#39;, &#39;spatstat.sparse&#39;, &#39;abind&#39;, &#39;tensor&#39;, &#39;goftest&#39;, &#39;deldir&#39;, &#39;polyclip&#39;, &#39;FNN&#39;, &#39;RSpectra&#39;, &#39;SummarizedExperiment&#39;, &#39;S4Vectors&#39;, &#39;BiocGenerics&#39;, &#39;GenomicRanges&#39;, &#39;DelayedArray&#39;, &#39;AnnotationHub&#39;, &#39;BiocFileCache&#39;, &#39;curl&#39;, &#39;rappdirs&#39;, &#39;limma&#39;, &#39;locfit&#39;, &#39;Rcpp&#39;, &#39;XML&#39;, &#39;AnnotationDbi&#39;, &#39;progress&#39;, &#39;httr&#39;, &#39;xml2&#39;, &#39;IRanges&#39;, &#39;BiocParallel&#39;, &#39;DelayedMatrixStats&#39;, &#39;HDF5Array&#39;, &#39;rhdf5&#39;, &#39;R.utils&#39;, &#39;dqrng&#39;, &#39;beachmat&#39;, &#39;Rhdf5lib&#39;, &#39;BH&#39;, &#39;ensembldb&#39;, &#39;GenomicFeatures&#39;, &#39;gridExtra&#39;, &#39;BiocNeighbors&#39;, &#39;ggbeeswarm&#39;, &#39;viridis&#39;, &#39;Rtsne&#39;, &#39;ggrepel&#39;, &#39;igraph&#39;, &#39;statmod&#39;, &#39;bluster&#39;, &#39;metapod&#39;, &#39;ScaledMatrix&#39;, &#39;irlba&#39;, &#39;rsvd&#39;, &#39;xgboost&#39;, &#39;cowplot&#39;, &#39;reshape2&#39;, &#39;fitdistrplus&#39;, &#39;future&#39;, &#39;future.apply&#39;, &#39;ggridges&#39;, &#39;ica&#39;, &#39;leiden&#39;, &#39;lmtest&#39;, &#39;matrixStats&#39;, &#39;miniUI&#39;, &#39;patchwork&#39;, &#39;pbapply&#39;, &#39;plotly&#39;, &#39;png&#39;, &#39;RANN&#39;, &#39;RcppAnnoy&#39;, &#39;reticulate&#39;, &#39;ROCR&#39;, &#39;scattermore&#39;, &#39;sctransform&#39;, &#39;SeuratObject&#39;, &#39;shiny&#39;, &#39;spatstat.core&#39;, &#39;spatstat.geom&#39;, &#39;uwot&#39;, &#39;RcppEigen&#39;, &#39;RcppProgress&#39;</code></pre>
<pre><code>## 
##   There are binary versions available but the source versions are later:
##          binary   source needs_compilation
## httpuv    1.6.3    1.6.4              TRUE
## BH     1.75.0-0 1.78.0-0             FALSE
## igraph    1.2.9   1.2.10              TRUE
## 
## 
## The downloaded binary packages are in
##  /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//RtmpDprmjP/downloaded_packages</code></pre>
<pre><code>## installing the source packages &#39;GenomeInfoDbData&#39;, &#39;httpuv&#39;, &#39;BH&#39;, &#39;igraph&#39;, &#39;scRNAseq&#39;</code></pre>
<pre><code>## Old packages: &#39;crayon&#39;, &#39;digest&#39;, &#39;glue&#39;, &#39;knitr&#39;, &#39;lattice&#39;, &#39;lifecycle&#39;,
##   &#39;mgcv&#39;, &#39;mime&#39;, &#39;nlme&#39;, &#39;pillar&#39;, &#39;rlang&#39;, &#39;rmarkdown&#39;, &#39;stringi&#39;, &#39;tibble&#39;,
##   &#39;tinytex&#39;, &#39;withr&#39;, &#39;xfun&#39;, &#39;Matrix&#39;</code></pre>
</div>
<div id="the-macosko-dataset" class="section level1">
<h1><span class="header-section-number">2</span> The Macosko dataset</h1>
<p>In this workshop session, we will analyze the single-cell RNA-seq dataset from the publication by Macosko <em>et al.</em>, Cell 161, 1202–1214 from 2015 <a href="https://doi.org/10.1016/j.cell.2015.05.002">(link)</a>. This is the manuscript in which the droplet scRNA-seq technology <strong>Drop-seq</strong> was introduced. Six years after the original publication, drop-seq is still one of the most commonly adopted scRNA-seq protocols, as evidenced by the large number of citations for Macosko <em>et al.</em> (4.303 citations at November 3, 2021).</p>
<p>The basic idea behind the Drop-seq protocol can be taken from the graphical abstract of the publication.</p>
<pre><code>knitr::include_graphics(&quot;macosko_graphicalAbstract.jpeg&quot;)</code></pre>
<p>The success of Drop-seq can be explained by the following advantageous features:</p>
<ul>
<li><p>The use of unique molecular identifiers (UMIs). By working with UMIs, one count corresponds to one observed mRNA molecule present in the cell. Thanks to the use of UMI barcodes, PCR artifacts are reduced.</p></li>
<li><p>Scalability: microfluidics technology allows for performing the library prep reactions inside nanodroplets, in which single cells may be contained. Library prep occurs across all droplets simultaneously.</p></li>
<li><p>Cost: the experiment costs around 6.5 cents (USD) per cell.</p></li>
<li><p>Speed: The very large dataset that we will be working with today was generated in an experiment that took only 4 days.</p></li>
</ul>
<p>In this particular experiment, Macosko <em>et al.</em> sequenced 49,300 cells from the mouse retina, identifying 39 transcriptionally distinct cell populations. The experiment was performed in 7 batches.</p>
</div>
<div id="data-availability" class="section level1">
<h1><span class="header-section-number">3</span> Data availability</h1>
<div id="sra" class="section level2">
<h2><span class="header-section-number">3.1</span> SRA</h2>
<p>The <a href="https://www.ncbi.nlm.nih.gov/sra">Sequence Read Archive (SRA)</a> is the largest publicly available repository of high-throughput sequencing data. The data are stored by the National Center for Biotechnology Information (NCBI) services and multiple cloud storage providers. From this website “raw” sequencing data can be retrieved. In practice, these are usually <code>.sra</code> files, which can be downloaded and converted into FASTQ files using functions from the <code>sratoolkit</code> software.</p>
<p>For our dataset, the FASTQ data can be retrieved from this <a href="https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA267857&amp;o=acc_s%3Aa">link</a>. The data are stored as one file per sequencing batch, with each file approximately 20Gb. As such, it will be unfeasible to download and process these FASTQ files in this practical session.</p>
<p>Instead, for demonstrative purposes, we have taken a subsample of the FASTQ file for the first sequencing batch for you to work with. On this subsample, we may perform all the tasks that we would have performed on the full dataset. The steps that are required for downloading and quantifying drop-seq data can be found in <a href="https://github.com/statOmics/singleCellCourse/blob/master/lab1_preprocessing/preprocessDropseq.sh">a Shell script on our companion GitHub repository</a>.</p>
</div>
<div id="geo" class="section level2">
<h2><span class="header-section-number">3.2</span> GEO</h2>
<p>The dataset from Macosko <em>et al.</em> was also uploaded by the authors on the Gene Expression Omnibus (GEO) platform under accession number <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63472">GSE63472</a>, from which raw and readily-processed data files may be retrieved, including:</p>
<ol style="list-style-type: decimal">
<li><p><em>GSE63472_RAW.tar</em>, a 90.6Gb object that contains the “raw data” for the experiment. In the scRNA-seq context, FASTQ files are often considered the raw data format.</p></li>
<li><p><em>GSE63472_P14Retina_merged_digital_expression.txt.gz</em>, a 50.7Mb matrix that stores the gene expression values for each cell. These values are integer counts, that did not undergo any type of preprocessing or normalization.</p></li>
<li><p><em>GSE63472_mm10_reference_metadata.tar.gz</em>, a 862.9Mb compressed folder containing information on the reference genome to which the scRNA-seq reads were aligned (see theory slides).</p></li>
<li><p><em>GSE63472_P14Retina_logDGE.txt.gz</em>, a 316.8Mb compressed text file, not clear what it contains (results from a differential gene expression analysis, but with log-transformation, so log-fold changes maybe?).</p></li>
</ol>
<p>As such, by downloading <em>GSE63472_P14Retina_merged_digital_expression.txt.gz</em>, we avoid re-quantifying the data, i.e., the translation from reads from the FASTQ files into a gene-level expression values for each cell.</p>
<p>One issue that often arises from data downloaded from GEO, is that there is no strict requirements for which data should be included in the upload by the authors. As such, from my personal experience, it can often be the case that important information like metadata are missing, or the content of the submitted files is unclear. Even if all the required data are available, as is the case here, we would still need to piece all the information together from different files and file formats before we can use them.</p>
</div>
<div id="experimenthub" class="section level2">
<h2><span class="header-section-number">3.3</span> ExperimentHub</h2>
<p>The Bioconductor <em>ExperimentHub</em> web resource, which can be accessed using the <a href="https://bioconductor.org/packages/release/bioc/html/ExperimentHub.html">ExperimentHub</a> R package, provides a central location where curated data from experiments, publications or training courses can be accessed. While it contains far less datasets than the SRA or GEO (4965 records to date), these datasets all follow the tidy data format of Bioconductor. Note that the ExperimentHub contains several types of data, like bulk and single-cell transcriptomics data, microarrays and more.</p>
<p>The Macosko dataset is available from ExperimentHub and can be accessed as follows:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="kw">library</span>(ExperimentHub)</span>
<span id="cb12-2"><a href="#cb12-2"></a>edb &lt;-<span class="st"> </span><span class="kw">ExperimentHub</span>()</span>
<span id="cb12-3"><a href="#cb12-3"></a></span>
<span id="cb12-4"><a href="#cb12-4"></a>edb[<span class="kw">grep</span>(<span class="st">&quot;Macosko&quot;</span>, edb<span class="op">$</span>title)] <span class="co"># find accession number (can be inefficient)</span></span>
<span id="cb12-5"><a href="#cb12-5"></a></span>
<span id="cb12-6"><a href="#cb12-6"></a>edb_counts &lt;-<span class="st"> </span>edb[[<span class="st">&quot;EH2690&quot;</span>]]</span>
<span id="cb12-7"><a href="#cb12-7"></a>edb_counts[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb12-8"><a href="#cb12-8"></a></span>
<span id="cb12-9"><a href="#cb12-9"></a>edb_coldata &lt;-<span class="st"> </span>edb[[<span class="st">&quot;EH2691&quot;</span>]]</span>
<span id="cb12-10"><a href="#cb12-10"></a>edb_coldata[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,]</span>
<span id="cb12-11"><a href="#cb12-11"></a></span>
<span id="cb12-12"><a href="#cb12-12"></a><span class="kw">rm</span>(edb, edb_counts, edb_coldata)</span></code></pre></div>
</div>
<div id="scrnaseq" class="section level2">
<h2><span class="header-section-number">3.4</span> scRNASeq</h2>
<p>In addition to ExperimentHub, Bioconductor provides the package <a href="https://bioconductor.org/packages/release/data/experiment/html/scRNAseq.html">scRNAseq</a>. This package provides an even more user-friendly client to access (only) scRNA-seq datasets from the ExperimentHub web resource. Data retrieved using the scRNAseq package are stored as user-friendly <code>SingleCellExperiment</code> objects, with the expression data, gene-level information, cell-level information and experiment metadata all in place in one data object. The scRNA-seq package currently holds 61 datasets, including the data from <em>Macosko et al.</em>:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="kw">library</span>(scRNAseq)</span>
<span id="cb13-2"><a href="#cb13-2"></a>scRNAseq<span class="op">::</span><span class="kw">MacoskoRetinaData</span>()</span></code></pre></div>
</div>
</div>
<div id="import-data" class="section level1">
<h1><span class="header-section-number">4</span> Import data</h1>
<p>The <code>scRNAseq</code> package provides convenient access to several datasets. See the <a href="http://bioconductor.org/packages/release/data/experiment/html/scRNAseq.html">package Bioconductor page</a> for more information.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="co"># Code below might ask you to create an ExperimentHub directory. </span></span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="co"># Type &#39;yes&#39; and hit Enter, to allow this.</span></span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(scRNAseq))</span>
<span id="cb14-4"><a href="#cb14-4"></a>sce &lt;-<span class="st"> </span><span class="kw">MacoskoRetinaData</span>()</span></code></pre></div>
</div>
<div id="a-singlecellexperiment-object" class="section level1">
<h1><span class="header-section-number">5</span> A <code>SingleCellExperiment</code> object</h1>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>sce</span></code></pre></div>
<div id="accessing-data-from-a-singlecellexperiment-object" class="section level2">
<h2><span class="header-section-number">5.1</span> Accessing data from a <code>SingleCellExperiment</code> object</h2>
<p>Please see <a href="http://bioconductor.org/books/release/OSCA/data-infrastructure.html">Figure 4.1 in OSCA</a> for an overview of a <code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="co"># Data: assays</span></span>
<span id="cb16-2"><a href="#cb16-2"></a><span class="kw">assays</span>(sce)</span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="kw">assays</span>(sce)<span class="op">$</span>counts[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb16-4"><a href="#cb16-4"></a></span>
<span id="cb16-5"><a href="#cb16-5"></a><span class="co"># Feature metadata: rowData</span></span>
<span id="cb16-6"><a href="#cb16-6"></a><span class="kw">rowData</span>(sce) <span class="co"># empty for now</span></span>
<span id="cb16-7"><a href="#cb16-7"></a></span>
<span id="cb16-8"><a href="#cb16-8"></a><span class="co"># Cell metadata: colData</span></span>
<span id="cb16-9"><a href="#cb16-9"></a><span class="kw">colData</span>(sce)</span>
<span id="cb16-10"><a href="#cb16-10"></a></span>
<span id="cb16-11"><a href="#cb16-11"></a><span class="co"># Reduced dimensions: reducedDims</span></span>
<span id="cb16-12"><a href="#cb16-12"></a><span class="kw">reducedDims</span>(sce) <span class="co"># empty for now</span></span></code></pre></div>
</div>
<div id="creating-a-new-singlecellexperiment-object" class="section level2">
<h2><span class="header-section-number">5.2</span> Creating a new <code>SingleCellExperiment</code> object</h2>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>sceNew &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> <span class="kw">assays</span>(sce)<span class="op">$</span>counts))</span>
<span id="cb17-2"><a href="#cb17-2"></a>sceNew</span>
<span id="cb17-3"><a href="#cb17-3"></a></span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="kw">rm</span>(sceNew)</span></code></pre></div>
</div>
<div id="storing-metadata-in-a-singlecellexperiment-object" class="section level2">
<h2><span class="header-section-number">5.3</span> Storing (meta)data in a <code>SingleCellExperiment</code> object</h2>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>fakeGeneNames &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sce))</span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="kw">rowData</span>(sce)<span class="op">$</span>fakeName &lt;-<span class="st"> </span>fakeGeneNames</span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="kw">head</span>(<span class="kw">rowData</span>(sce))</span>
<span id="cb18-4"><a href="#cb18-4"></a><span class="co"># Remove again by setting to NULL</span></span>
<span id="cb18-5"><a href="#cb18-5"></a><span class="kw">rowData</span>(sce)<span class="op">$</span>fakeName &lt;-<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb18-6"><a href="#cb18-6"></a></span>
<span id="cb18-7"><a href="#cb18-7"></a><span class="kw">assays</span>(sce)<span class="op">$</span>logCounts &lt;-<span class="st"> </span><span class="kw">log1p</span>(<span class="kw">assays</span>(sce)<span class="op">$</span>counts)</span>
<span id="cb18-8"><a href="#cb18-8"></a><span class="kw">assays</span>(sce)</span>
<span id="cb18-9"><a href="#cb18-9"></a><span class="kw">assays</span>(sce)<span class="op">$</span>logCounts[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb18-10"><a href="#cb18-10"></a><span class="kw">assays</span>(sce)<span class="op">$</span>logCounts &lt;-<span class="st"> </span><span class="ot">NULL</span></span></code></pre></div>
</div>
</div>
<div id="obtaining-and-including-rowdata" class="section level1">
<h1><span class="header-section-number">6</span> Obtaining and including rowData</h1>
<p>The <code>rowData</code> slot of a <code>SingleCellExperiment</code> object allows for storing information on the features, i.e. the genes, in a dataset. In our object, the <code>rowData</code> slot is empty.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="kw">rowData</span>(sce)</span></code></pre></div>
<p>As such, the only information we have on the genes are their names, which can be retrieved as the <code>rownames</code> of the expression matrix.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="kw">head</span>(<span class="kw">rownames</span>(sce))</span></code></pre></div>
<p>These are the gene names (symbols). Note that it may be useful to include additional information in the <code>rowData</code> slot. For instance, we may want to store:</p>
<ul>
<li>Unambiguous gene identifiers (e.g. from ENSEMBL)</li>
<li>On which chromosome the gene is located</li>
<li>Gene length (genomic start position and end position)</li>
<li>Others…</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="kw">library</span>(<span class="st">&quot;biomaRt&quot;</span>)</span>
<span id="cb21-2"><a href="#cb21-2"></a></span>
<span id="cb21-3"><a href="#cb21-3"></a>ensembl75 &lt;-<span class="st"> </span><span class="kw">useEnsembl</span>(<span class="dt">biomart =</span> <span class="st">&#39;genes&#39;</span>, </span>
<span id="cb21-4"><a href="#cb21-4"></a>                        <span class="dt">dataset =</span> <span class="st">&#39;mmusculus_gene_ensembl&#39;</span>,</span>
<span id="cb21-5"><a href="#cb21-5"></a>                        <span class="dt">version =</span> <span class="dv">75</span>)</span>
<span id="cb21-6"><a href="#cb21-6"></a></span>
<span id="cb21-7"><a href="#cb21-7"></a><span class="kw">head</span>(<span class="kw">listAttributes</span>(ensembl75)) <span class="co"># potential info to extract</span></span>
<span id="cb21-8"><a href="#cb21-8"></a></span>
<span id="cb21-9"><a href="#cb21-9"></a>geneInfo &lt;-<span class="st"> </span><span class="kw">getBM</span>(<span class="dt">attributes =</span> <span class="kw">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="co"># ENSEMBL unambiguous identifier</span></span>
<span id="cb21-10"><a href="#cb21-10"></a>                                 <span class="st">&quot;mgi_symbol&quot;</span>, <span class="co"># Gene symbol (to link with SCE rownames),</span></span>
<span id="cb21-11"><a href="#cb21-11"></a>                                 <span class="st">&quot;chromosome_name&quot;</span>, <span class="co"># On which chromosome</span></span>
<span id="cb21-12"><a href="#cb21-12"></a>                                 <span class="st">&quot;start_position&quot;</span>, <span class="co"># Start position</span></span>
<span id="cb21-13"><a href="#cb21-13"></a>                                 <span class="st">&quot;end_position&quot;</span>),<span class="co"># End position</span></span>
<span id="cb21-14"><a href="#cb21-14"></a>                  <span class="dt">mart =</span> ensembl75)</span>
<span id="cb21-15"><a href="#cb21-15"></a></span>
<span id="cb21-16"><a href="#cb21-16"></a><span class="kw">head</span>(geneInfo)</span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>geneInfo<span class="op">$</span>mgi_symbol_upper &lt;-<span class="st"> </span><span class="kw">toupper</span>(geneInfo<span class="op">$</span>mgi_symbol) </span>
<span id="cb22-2"><a href="#cb22-2"></a><span class="co"># match between gene info and rownames</span></span>
<span id="cb22-3"><a href="#cb22-3"></a></span>
<span id="cb22-4"><a href="#cb22-4"></a><span class="kw">sum</span>(<span class="kw">rownames</span>(sce) <span class="op">%in%</span><span class="st"> </span>geneInfo<span class="op">$</span>mgi_symbol_upper)</span>
<span id="cb22-5"><a href="#cb22-5"></a><span class="kw">sum</span>(<span class="op">!</span><span class="kw">rownames</span>(sce) <span class="op">%in%</span><span class="st"> </span>geneInfo<span class="op">$</span>mgi_symbol_upper) <span class="co"># lost in conversion :(</span></span>
<span id="cb22-6"><a href="#cb22-6"></a></span>
<span id="cb22-7"><a href="#cb22-7"></a><span class="kw">rowData</span>(sce) &lt;-<span class="st"> </span>geneInfo[<span class="kw">match</span>(<span class="kw">rownames</span>(sce),geneInfo<span class="op">$</span>mgi_symbol_upper),]</span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="kw">head</span>(<span class="kw">rowData</span>(sce))</span></code></pre></div>
</div>
<div id="filtering-non-informative-genes" class="section level1">
<h1><span class="header-section-number">7</span> Filtering non-informative genes</h1>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="kw">library</span>(edgeR)</span>
<span id="cb24-2"><a href="#cb24-2"></a></span>
<span id="cb24-3"><a href="#cb24-3"></a><span class="co"># A very simple strategy: remove all genes that are expressed in less than 10</span></span>
<span id="cb24-4"><a href="#cb24-4"></a><span class="co"># out of 49300 cells -&gt; note that this a very lenient filtering criterium</span></span>
<span id="cb24-5"><a href="#cb24-5"></a>keep &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">assays</span>(sce)<span class="op">$</span>counts <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span></span>
<span id="cb24-6"><a href="#cb24-6"></a><span class="kw">table</span>(keep)</span>
<span id="cb24-7"><a href="#cb24-7"></a></span>
<span id="cb24-8"><a href="#cb24-8"></a>sce &lt;-<span class="st"> </span>sce[keep,]</span></code></pre></div>
<p>Note that dedicated functions for filtering out lowly expressed genes exist. One such function is the <code>filterByExpr</code> function of the edgeR R package. In brief, the strategy keeps genes that have at least “min.count” reads in a worthwhile number samples.</p>
<p>More precisely, the filtering keeps genes that have count-per-million (CPM) above k in n samples.</p>
<p>k is determined by the min.count argument to the function, and by the sample library sizes.</p>
<p>n is determined by the design matrix. n can be seen as the smallest group sample size. A <code>group</code> of samples/cells can be defined as cells that are more similar to one another, e.g., from the same sequencing batch, the same patient…. Here we could also use the cluster assignment (cell type) for each cell as the grouping variable; <em>note however, that this usually is not available until a</em> <em>later stage in the analysis pipeline</em> (i.e., after dimension reduction and clustering, topics we will cover next session.)</p>
<p>If all the group sizes are larger than the <code>large.n</code> argument of the filterByExpr function, which defaults to 10, then n will be taken as <code>min.prop</code>* the the number of samples/cells in the smallest group.</p>
<p>Note that all the group sizes will often be larger than the <code>large.n</code> in case of single-cell data.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="co"># Slow (more than 1min) -&gt; do not run</span></span>
<span id="cb25-2"><a href="#cb25-2"></a><span class="co"># keep2 &lt;- filterByExpr(sce,</span></span>
<span id="cb25-3"><a href="#cb25-3"></a><span class="co">#                       group = sce$cluster,</span></span>
<span id="cb25-4"><a href="#cb25-4"></a><span class="co">#                       min.count = 1,</span></span>
<span id="cb25-5"><a href="#cb25-5"></a><span class="co">#                       min.total.count = 15,</span></span>
<span id="cb25-6"><a href="#cb25-6"></a><span class="co">#                       min.prop = 0.2)</span></span>
<span id="cb25-7"><a href="#cb25-7"></a><span class="co"># table(keep2)</span></span></code></pre></div>
</div>
<div id="quality-control" class="section level1">
<h1><span class="header-section-number">8</span> Quality control</h1>
<p>In quality control (QC), we check the quality of our dataset. In particular, we investigate undesirable oddities, such as low-quality cells, empty droplets or doublets.</p>
<div id="identifying-and-removing-low-quality-cells" class="section level2">
<h2><span class="header-section-number">8.1</span> Identifying and removing low-quality cells</h2>
<p>There are several distinguishing features of low-quality cells that can be used in order to identify them. <a href="http://bioconductor.org/books/3.14/OSCA.basic/quality-control.html">As described in the OSCA book</a> book):</p>
<ol style="list-style-type: decimal">
<li><p>The library size is defined as the total sum of counts across all relevant features for each cell. Here, we will consider the relevant features to be the endogenous genes. Cells with small library sizes are of low quality as the RNA has been lost at some point during library preparation, either due to cell lysis or inefficient cDNA capture and amplification.</p></li>
<li><p>The number of expressed features in each cell is defined as the number of endogenous genes with non-zero counts for that cell. Any cell with very few expressed genes is likely to be of poor quality as the diverse transcript population has not been successfully captured.</p></li>
<li><p>We sometimes have spike-in (ERCC) transcripts available. The proportion of reads mapped to spike-in transcripts is calculated relative to the total count across all features (including spike-ins) for each cell. As the same amount of spike-in RNA should have been added to each cell, any enrichment in spike-in counts is symptomatic of loss of endogenous RNA.</p></li>
<li><p>In the absence of spike-in transcripts, the proportion of reads mapped to genes in the mitochondrial genome can be used. High proportions are indicative of poor-quality cells (Islam et al. 2014; Ilicic et al. 2016), presumably because of loss of cytoplasmic RNA from perforated cells. The reasoning is that, in the presence of modest damage, the holes in the cell membrane permit efflux of individual transcript molecules but are too small to allow mitochondria to escape, leading to a relative enrichment of mitochondrial transcripts. For single-nuclei RNA-seq experiments, high proportions are also useful as they can mark cells where the cytoplasm has not been successfully stripped.</p></li>
</ol>
</div>
<div id="calculate-qc-variables" class="section level2">
<h2><span class="header-section-number">8.2</span> Calculate QC variables</h2>
<p>This function calculates useful QC metrics for identification and removal of potentially problematic cells. Default per-cell metrics are the sum of counts (i.e., the library size) and the number of detected features. The percentage of counts in the top features also provides a measure of library complexity.</p>
<p>If subsets is specified, these statistics are also computed for each subset of features. This is useful for investigating gene sets of interest, e.g., mitochondrial genes.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="kw">library</span>(scater)</span>
<span id="cb26-2"><a href="#cb26-2"></a></span>
<span id="cb26-3"><a href="#cb26-3"></a><span class="co"># check ERCC spike-in transcripts</span></span>
<span id="cb26-4"><a href="#cb26-4"></a><span class="kw">sum</span>(<span class="kw">grepl</span>(<span class="st">&quot;^ERCC-&quot;</span>, <span class="kw">rownames</span>(sce))) <span class="co"># no spike-in transcripts available</span></span>
<span id="cb26-5"><a href="#cb26-5"></a></span>
<span id="cb26-6"><a href="#cb26-6"></a><span class="co"># check mitochondrial genes</span></span>
<span id="cb26-7"><a href="#cb26-7"></a><span class="kw">sum</span>(<span class="kw">rowData</span>(sce)<span class="op">$</span>chromosome_name<span class="op">==</span><span class="st">&quot;MT&quot;</span>,<span class="dt">na.rm =</span> <span class="ot">TRUE</span>) <span class="co"># 28 mitochondrial genes</span></span>
<span id="cb26-8"><a href="#cb26-8"></a><span class="kw">sum</span>(<span class="kw">grepl</span>(<span class="st">&quot;^MT-&quot;</span>, <span class="kw">rownames</span>(sce))) <span class="co"># alternatively</span></span>
<span id="cb26-9"><a href="#cb26-9"></a>is.mito &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;^MT-&quot;</span>, <span class="kw">rownames</span>(sce))</span>
<span id="cb26-10"><a href="#cb26-10"></a></span>
<span id="cb26-11"><a href="#cb26-11"></a><span class="co">## calculate QC metrics</span></span>
<span id="cb26-12"><a href="#cb26-12"></a>df &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce, <span class="dt">subsets=</span><span class="kw">list</span>(<span class="dt">Mito=</span>is.mito))</span>
<span id="cb26-13"><a href="#cb26-13"></a><span class="kw">head</span>(df)</span>
<span id="cb26-14"><a href="#cb26-14"></a></span>
<span id="cb26-15"><a href="#cb26-15"></a><span class="co"># add QC variables to sce object</span></span>
<span id="cb26-16"><a href="#cb26-16"></a><span class="kw">colData</span>(sce) &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">colData</span>(sce), df)</span>
<span id="cb26-17"><a href="#cb26-17"></a></span>
<span id="cb26-18"><a href="#cb26-18"></a><span class="co"># the QC variables have now been added to the colData of our SCE object.</span></span>
<span id="cb26-19"><a href="#cb26-19"></a><span class="kw">head</span>(<span class="kw">colData</span>(sce))</span></code></pre></div>
</div>
<div id="exploratory-data-analysis" class="section level2">
<h2><span class="header-section-number">8.3</span> Exploratory data analysis</h2>
<p>In the figure below, we see that several cells have a very low number of expressed genes, and where most of the molecules are derived from mitochondrial genes. This indicates likely damaged cells, presumably because of loss of cytoplasmic RNA from perforated cells, so we should remove these for the downstream analysis.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="co"># Number of genes vs library size</span></span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="kw">plotColData</span>(sce, <span class="dt">x =</span> <span class="st">&quot;sum&quot;</span>, <span class="dt">y=</span><span class="st">&quot;detected&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;cluster&quot;</span>) </span>
<span id="cb27-3"><a href="#cb27-3"></a></span>
<span id="cb27-4"><a href="#cb27-4"></a><span class="co"># Mitochondrial genes</span></span>
<span id="cb27-5"><a href="#cb27-5"></a><span class="kw">plotColData</span>(sce, <span class="dt">x =</span> <span class="st">&quot;detected&quot;</span>, <span class="dt">y=</span><span class="st">&quot;subsets_Mito_percent&quot;</span>)</span></code></pre></div>
</div>
<div id="qc-using-adaptive-thresholds" class="section level2">
<h2><span class="header-section-number">8.4</span> QC using adaptive thresholds</h2>
<p>Below, we remove cells that are outlying with respect to</p>
<ol style="list-style-type: decimal">
<li>A low sequencing depth (number of UMIs);</li>
<li>A low number of genes detected;</li>
<li>A high percentage of reads from mitochondrial genes.</li>
</ol>
<p>Here we will remove cells for QC based on <strong>adaptive thresholds</strong> related to the three points from above. Adaptive trhesholds are used as opposed to <strong>fixed thresholds</strong>.</p>
<p>With fixed thresholds, we use fixed cut-off values for each cell to pass QC, e.g., we might consider cells to be low quality if they have library sizes below 100,000 reads; express fewer than 5,000 genes; have spike-in proportions above 10%; or have mitochondrial proportions above 10%.</p>
<p>With adaptive thresholds, we assume that most of the dataset consists of high-quality cells. We then identify cells that are outliers for the various QC metrics, based on the median absolute deviation (MAD) from the median value of each metric across all cells. By default, we consider a value to be anoutlier if it is more than 3 MADs from the median in the “problematic” direction. This is loosely motivated by the fact that such a filter will retain 99% of non-outlier values that follow a normal distribution. We demonstrate adopting adaptive thresholds on the Macosko dataset:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>lowLib &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(df<span class="op">$</span>sum, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">log=</span><span class="ot">TRUE</span>)</span>
<span id="cb28-2"><a href="#cb28-2"></a>lowFeatures &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(df<span class="op">$</span>detected, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">log=</span><span class="ot">TRUE</span>)</span>
<span id="cb28-3"><a href="#cb28-3"></a>highMito &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(df<span class="op">$</span>subsets_Mito_percent, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>)</span>
<span id="cb28-4"><a href="#cb28-4"></a></span>
<span id="cb28-5"><a href="#cb28-5"></a><span class="kw">table</span>(lowLib)</span>
<span id="cb28-6"><a href="#cb28-6"></a><span class="kw">table</span>(lowFeatures)</span>
<span id="cb28-7"><a href="#cb28-7"></a><span class="kw">table</span>(highMito)</span>
<span id="cb28-8"><a href="#cb28-8"></a></span>
<span id="cb28-9"><a href="#cb28-9"></a>discardCells &lt;-<span class="st"> </span>(lowLib <span class="op">|</span><span class="st"> </span>lowFeatures <span class="op">|</span><span class="st"> </span>highMito)</span>
<span id="cb28-10"><a href="#cb28-10"></a><span class="kw">table</span>(discardCells)</span>
<span id="cb28-11"><a href="#cb28-11"></a><span class="kw">colData</span>(sce)<span class="op">$</span>discardCells &lt;-<span class="st"> </span>discardCells</span>
<span id="cb28-12"><a href="#cb28-12"></a></span>
<span id="cb28-13"><a href="#cb28-13"></a><span class="co"># visualize cells to be removed</span></span>
<span id="cb28-14"><a href="#cb28-14"></a><span class="kw">plotColData</span>(sce, <span class="dt">x =</span> <span class="st">&quot;sum&quot;</span>, <span class="dt">y=</span><span class="st">&quot;detected&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;discardCells&quot;</span>)</span>
<span id="cb28-15"><a href="#cb28-15"></a><span class="kw">plotColData</span>(sce, <span class="dt">x =</span> <span class="st">&quot;detected&quot;</span>, <span class="dt">y=</span><span class="st">&quot;subsets_Mito_percent&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;discardCells&quot;</span>)</span></code></pre></div>
<p>We removed a total of <span class="math inline">\(3423\)</span> cells, most of which because of an outlyingly high percentage of reads from mitochondrial genes.</p>
</div>
<div id="identifying-and-removing-empty-droplets" class="section level2">
<h2><span class="header-section-number">8.5</span> Identifying and removing empty droplets</h2>
<p>Note that the removal of cells with low sequencing depth using the adaptive threshold procedure above is a way of removing empty droplets. Other approaches are possible, e.g., removing cells by statistical testing using <code>emtpyDrops</code>. This does require us to specify a lower bound on the total number of UMIs, below which all cells are considered to correspond to empty droplets. This lower bound may not be trivial to derive, but the <code>barcodeRanks</code> function can be useful to identify an elbow/knee point.</p>
<p>In brief, the steps taken by the <code>emtpyDrops</code> function can be summarized as follows:</p>
<ol style="list-style-type: decimal">
<li><p>Define threshold T of total UMI counts (e.g. with the help of the <code>barcodeRanks</code> function), below which cells may be considered to be from empty droplets. Call this set of cells E.</p></li>
<li><p>Define <span class="math inline">\(A_g\)</span> as the total gene expression across all cells in E.</p></li>
<li><p>Define <span class="math inline">\(pi_g\)</span> as the relative contribution of gene g to the ambient profile.</p></li>
<li><p>Calculate p-value for each cell to have a transcriptional profile similar to the ambient solution. Intuitively, a p-value below the requested alpha level would correspond to a cell for which the observed count profile strongly deviates from the count profile observed in the cells with a library size below threshold T, i.e., a non-empty droplet.</p></li>
</ol>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="kw">library</span>(DropletUtils)</span>
<span id="cb29-2"><a href="#cb29-2"></a>bcrank &lt;-<span class="st"> </span><span class="kw">barcodeRanks</span>(<span class="kw">counts</span>(sce))</span>
<span id="cb29-3"><a href="#cb29-3"></a></span>
<span id="cb29-4"><a href="#cb29-4"></a><span class="co"># Only showing unique points for plotting speed. Duplicated ranks are a </span></span>
<span id="cb29-5"><a href="#cb29-5"></a><span class="co"># consequence of ties in the ranks, i.e., when cells have an equal library size.</span></span>
<span id="cb29-6"><a href="#cb29-6"></a><span class="kw">sum</span>(<span class="kw">duplicated</span>(bcrank<span class="op">$</span>rank))</span>
<span id="cb29-7"><a href="#cb29-7"></a>uniq &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">duplicated</span>(bcrank<span class="op">$</span>rank)</span>
<span id="cb29-8"><a href="#cb29-8"></a></span>
<span id="cb29-9"><a href="#cb29-9"></a><span class="kw">plot</span>(bcrank<span class="op">$</span>rank[uniq], bcrank<span class="op">$</span>total[uniq], <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>,</span>
<span id="cb29-10"><a href="#cb29-10"></a>    <span class="dt">xlab=</span><span class="st">&quot;Rank&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Total UMI count&quot;</span>, <span class="dt">cex.lab=</span><span class="fl">1.2</span>)</span>
<span id="cb29-11"><a href="#cb29-11"></a><span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">metadata</span>(bcrank)<span class="op">$</span>inflection, <span class="dt">col=</span><span class="st">&quot;darkgreen&quot;</span>, <span class="dt">lty=</span><span class="dv">2</span>)</span>
<span id="cb29-12"><a href="#cb29-12"></a><span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">metadata</span>(bcrank)<span class="op">$</span>knee, <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">lty=</span><span class="dv">2</span>)</span>
<span id="cb29-13"><a href="#cb29-13"></a><span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">350</span>, <span class="dt">col=</span><span class="st">&quot;orange&quot;</span>, <span class="dt">lty=</span><span class="dv">2</span>) <span class="co"># picked visually myself</span></span>
<span id="cb29-14"><a href="#cb29-14"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, </span>
<span id="cb29-15"><a href="#cb29-15"></a>       <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;Inflection&quot;</span>, <span class="st">&quot;Knee&quot;</span>, <span class="st">&quot;Empirical knee point&quot;</span>), </span>
<span id="cb29-16"><a href="#cb29-16"></a>       <span class="dt">col=</span><span class="kw">c</span>(<span class="st">&quot;darkgreen&quot;</span>, <span class="st">&quot;dodgerblue&quot;</span>, <span class="st">&quot;orange&quot;</span>), </span>
<span id="cb29-17"><a href="#cb29-17"></a>       <span class="dt">lty=</span><span class="dv">2</span>, </span>
<span id="cb29-18"><a href="#cb29-18"></a>       <span class="dt">cex=</span><span class="fl">1.2</span>)</span>
<span id="cb29-19"><a href="#cb29-19"></a></span>
<span id="cb29-20"><a href="#cb29-20"></a><span class="kw">set.seed</span>(<span class="dv">100</span>)</span>
<span id="cb29-21"><a href="#cb29-21"></a>limit &lt;-<span class="st"> </span><span class="dv">350</span>   </span>
<span id="cb29-22"><a href="#cb29-22"></a>all.out &lt;-<span class="st"> </span><span class="kw">emptyDrops</span>(<span class="kw">counts</span>(sce), <span class="dt">lower=</span>limit, <span class="dt">test.ambient=</span><span class="ot">TRUE</span>)</span>
<span id="cb29-23"><a href="#cb29-23"></a><span class="co"># p-values for cells with total UMI count under the lower bound.</span></span>
<span id="cb29-24"><a href="#cb29-24"></a><span class="kw">hist</span>(all.out<span class="op">$</span>PValue[all.out<span class="op">$</span>Total <span class="op">&lt;=</span><span class="st"> </span>limit <span class="op">&amp;</span><span class="st"> </span>all.out<span class="op">$</span>Total <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>],</span>
<span id="cb29-25"><a href="#cb29-25"></a>    <span class="dt">xlab=</span><span class="st">&quot;P-value&quot;</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>, <span class="dt">col=</span><span class="st">&quot;grey80&quot;</span>, <span class="dt">breaks=</span><span class="dv">20</span>)</span>
<span id="cb29-26"><a href="#cb29-26"></a></span>
<span id="cb29-27"><a href="#cb29-27"></a><span class="co"># but note that it would remove a very high number of cells</span></span>
<span id="cb29-28"><a href="#cb29-28"></a><span class="kw">length</span>(<span class="kw">which</span>(all.out<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.01</span>)) <span class="co"># retained</span></span>
<span id="cb29-29"><a href="#cb29-29"></a></span>
<span id="cb29-30"><a href="#cb29-30"></a><span class="co"># so we stick to the more lenient adaptive filtering strategy</span></span>
<span id="cb29-31"><a href="#cb29-31"></a><span class="co"># remove cells identified using adaptive thresholds</span></span>
<span id="cb29-32"><a href="#cb29-32"></a>sce &lt;-<span class="st"> </span>sce[, <span class="op">!</span><span class="kw">colData</span>(sce)<span class="op">$</span>discardCells]</span></code></pre></div>
</div>
<div id="identifying-and-removing-doublets" class="section level2">
<h2><span class="header-section-number">8.6</span> Identifying and removing doublets</h2>
<p>We will use <a href="https://bioconductor.org/packages/3.14/bioc/html/scDblFinder.html">scDblFinder</a> to detect doublet cells.</p>
<p>As discussed in the theory session of last week, the steps taken by <code>scDblFinder</code> can be summarized as follows:</p>
<ol style="list-style-type: decimal">
<li><p>Perform principal components analysis (PCA) on log-normalized expression counts. This allows for projecting each cell in the dataset into a 2D space (for more details on PCA, see next weeks session).</p></li>
<li><p>Randomly select two cells, sum their counts and normalize, and project into PCA space from step 1. In other words, artificially generate doublets and see where they are located in the 2D space.</p></li>
<li><p>Repeat step 2 many times (generate many artificial doublets).</p></li>
<li><p>Generate neighbor network in the 2D space. The network is then used to estimate a number of characteristics for each cell, in particular the proportion f artificial doublets among the nearest neighbors.</p></li>
<li><p>Use this information, along with other predictors, to train a classifier (gradient boosted tree) that allows for distinguishing doublets from singlets.</p></li>
</ol>
<p>Note: only classifies for identifying doublets for which the two cells are from different cell type clusters.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="co">## perform doublet detection</span></span>
<span id="cb30-2"><a href="#cb30-2"></a><span class="kw">library</span>(scDblFinder)</span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="kw">set.seed</span>(<span class="dv">211103</span>)</span>
<span id="cb31-2"><a href="#cb31-2"></a>sampleID &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>(<span class="kw">strsplit</span>(<span class="kw">colData</span>(sce)<span class="op">$</span>cell.id, <span class="dt">split=</span><span class="st">&quot;_&quot;</span>), <span class="st">&quot;[[&quot;</span>, <span class="dv">1</span>))</span>
<span id="cb31-3"><a href="#cb31-3"></a><span class="kw">table</span>(sampleID)</span>
<span id="cb31-4"><a href="#cb31-4"></a>sce &lt;-<span class="st"> </span><span class="kw">scDblFinder</span>(sce, </span>
<span id="cb31-5"><a href="#cb31-5"></a>                   <span class="dt">returnType=</span><span class="st">&quot;table&quot;</span>,</span>
<span id="cb31-6"><a href="#cb31-6"></a>                   <span class="dt">samples =</span> <span class="kw">factor</span>(sampleID))</span>
<span id="cb31-7"><a href="#cb31-7"></a><span class="kw">table</span>(sce<span class="op">$</span>scDblFinder.class)</span>
<span id="cb31-8"><a href="#cb31-8"></a></span>
<span id="cb31-9"><a href="#cb31-9"></a></span>
<span id="cb31-10"><a href="#cb31-10"></a><span class="co">## visualize these scores</span></span>
<span id="cb31-11"><a href="#cb31-11"></a><span class="co">## explore doublet score wrt original cluster labels</span></span>
<span id="cb31-12"><a href="#cb31-12"></a><span class="kw">boxplot</span>(<span class="kw">log1p</span>(sce<span class="op">$</span>scDblFinder.score) <span class="op">~</span><span class="st"> </span><span class="kw">factor</span>(<span class="kw">colData</span>(sce)<span class="op">$</span>cluster, <span class="dt">exclude=</span><span class="ot">NULL</span>))</span>
<span id="cb31-13"><a href="#cb31-13"></a></span>
<span id="cb31-14"><a href="#cb31-14"></a>tab &lt;-<span class="st"> </span><span class="kw">table</span>(sce<span class="op">$</span>scDblFinder.class, sce<span class="op">$</span>cluster, </span>
<span id="cb31-15"><a href="#cb31-15"></a>      <span class="dt">exclude=</span><span class="ot">NULL</span>)</span>
<span id="cb31-16"><a href="#cb31-16"></a>tab</span>
<span id="cb31-17"><a href="#cb31-17"></a><span class="kw">t</span>(<span class="kw">t</span>(tab) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(tab))</span>
<span id="cb31-18"><a href="#cb31-18"></a></span>
<span id="cb31-19"><a href="#cb31-19"></a><span class="kw">barplot</span>(<span class="kw">t</span>(<span class="kw">t</span>(tab) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(tab))[<span class="dv">2</span>,],</span>
<span id="cb31-20"><a href="#cb31-20"></a>        <span class="dt">xlab =</span> <span class="st">&quot;Cluster&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Fraction of doublets&quot;</span>)</span>
<span id="cb31-21"><a href="#cb31-21"></a></span>
<span id="cb31-22"><a href="#cb31-22"></a><span class="kw">range</span>(sce<span class="op">$</span>scDblFinder.score[sce<span class="op">$</span>scDblFinder.class  <span class="op">==</span><span class="st"> &quot;doublet&quot;</span> <span class="op">&amp;</span><span class="st"> </span>sampleID <span class="op">==</span><span class="st"> &quot;r1&quot;</span>])</span>
<span id="cb31-23"><a href="#cb31-23"></a><span class="kw">range</span>(sce<span class="op">$</span>scDblFinder.score[sce<span class="op">$</span>scDblFinder.class  <span class="op">==</span><span class="st"> &quot;singlet&quot;</span> <span class="op">&amp;</span><span class="st"> </span>sampleID <span class="op">==</span><span class="st"> &quot;r1&quot;</span>])</span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="co"># remove doublets</span></span>
<span id="cb32-2"><a href="#cb32-2"></a>sce &lt;-<span class="st"> </span>sce[,<span class="op">!</span>sce<span class="op">$</span>scDblFinder.class <span class="op">==</span><span class="st"> &quot;doublet&quot;</span>]</span></code></pre></div>
</div>
</div>
<div id="normalization" class="section level1">
<h1><span class="header-section-number">9</span> Normalization</h1>
<p>Normalization aims to remove technical effects such as sequencing depth so that comparisons between cells are not confounded by them. The most commonly used normalization methods methods use scaling, where a scaling factor (also called size factor, normalization factor) is estimated for each cell. These scaling factors (e.g., the effective library size) can be included as an offset to downstream modeling procedures. This effectively allows for performing inference on the relative abundance of a gene in a cell, after accounting for library size and RNA composition differences between cells, which is much more relevant than comparing raw counts</p>
<p>Note that we always prefer the use of offsets over transformation of the data. In brief, such transformation would distort the mean-variance relationship in the data. For more details, we refer to the document that was discussed in theory session of last week (<a href="https://statomics.github.io/SGA21/sequencing_scalingNormalization.html">link</a>)</p>
<p>For normalization, the size factors <span class="math inline">\(s_i\)</span> computed here are simply scaled library sizes: <span class="math display">\[ N_i = \sum_g Y_{gi} \]</span> <span class="math display">\[ s_i = N_i / \bar{N}_i \]</span></p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a>sce &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce)</span>
<span id="cb33-2"><a href="#cb33-2"></a></span>
<span id="cb33-3"><a href="#cb33-3"></a><span class="co"># note we also returned log counts: see the additional logcounts assay.</span></span>
<span id="cb33-4"><a href="#cb33-4"></a>sce</span>
<span id="cb33-5"><a href="#cb33-5"></a></span>
<span id="cb33-6"><a href="#cb33-6"></a><span class="co"># you can extract size factors using</span></span>
<span id="cb33-7"><a href="#cb33-7"></a>sf &lt;-<span class="st"> </span><span class="kw">librarySizeFactors</span>(sce)</span>
<span id="cb33-8"><a href="#cb33-8"></a><span class="kw">mean</span>(sf) <span class="co"># equal to 1 due to scaling.</span></span>
<span id="cb33-9"><a href="#cb33-9"></a><span class="kw">plot</span>(<span class="dt">x=</span> <span class="kw">log</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(sce)<span class="op">$</span>counts)), </span>
<span id="cb33-10"><a href="#cb33-10"></a>     <span class="dt">y=</span>sf)</span></code></pre></div>
<p>From the OSCA book: Alternatively, we may use more sophisticated approaches for variance stabilizing transformations in genomics data, e.g., <code>DESeq2</code> or <code>sctransform</code>. These aim to remove the mean-variance trend more effectively than the simpler transformations mentioned above, though it could be argued whether this is actually desirable. For low-coverage scRNA-seq data, there will always be a mean-variance trend under any transformation, for the simple reason that the variance must be zero when the mean count is zero. These methods also face the challenge of removing the mean-variance trend while preserving the interesting component of variation, i.e., the log-fold changes between subpopulations; this may or may not be done adequately, depending on the aggressiveness of the algorithm.</p>
<p>In practice, the log-transformation is a good default choice due to its simplicity and interpretability, and is what we will be using for all downstream analyses.</p>
<hr />
<p>— end lab session 1 —</p>
<hr />
<p>We have created a <a href="https://www.dropbox.com/sh/3yc9065ri76b2e7/AAAr0z9ZGp3APZ-reH6I0DGQa?dl=0">DropBox folder</a> containing convenience files, e.g., a preprocessed <code>SingleCellExperiment</code> that is the end result of lab session 1. You may want to downlaod them and import them in the code chunk below.</p>
</div>
<div id="normalization-continued" class="section level1">
<h1><span class="header-section-number">10</span> Normalization (continued)</h1>
<p>Normalization is necessary to correct for several sources of technical variation:</p>
<ul>
<li><strong>Differences in sequencing depth</strong> between samples. Some samples get sequenced deeper in the sense that they consist of more (mapped) reads and therefore can be considered to contain a higher amount of information, which we should be taking into account. In addition, if a sample is sequenced deeper, it is natural that the counts for each gene will be higher, jeopardizing a direct comparison of the expression counts.</li>
<li><strong>Differences in RNA population composition</strong> between samples. As an extreme example, suppose that two samples have been sequenced to the exact same depth. One sample is contaminated and has a very high concentration of the contaminant cDNA being sequenced, but otherwise the two samples are identical. Since the contaminant will be taking up a significant proportion of the reads being sequenced, the counts will not be directly comparable between the samples. Hence, we may also want to correct for differences in the composition of the RNA population of the samples (see <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf">edgeR manual chapter 2.8</a>).</li>
<li><strong>Other technical variation</strong> such as sample-specific GC-content or transcript length effects may also be accounted for.</li>
</ul>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a><span class="kw">table</span>(sce<span class="op">$</span>cluster)</span></code></pre></div>
<p>Below, we will visualize the normalization occurring between two cells of the same cell type (which could be considered technical repeats):</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a>select &lt;-<span class="st"> </span>sce<span class="op">$</span>cluster <span class="op">==</span><span class="st"> &quot;1&quot;</span></span>
<span id="cb35-2"><a href="#cb35-2"></a>select[<span class="kw">is.na</span>(select)] &lt;-<span class="st"> </span><span class="ot">FALSE</span></span>
<span id="cb35-3"><a href="#cb35-3"></a>cs &lt;-<span class="st"> </span><span class="kw">colSums</span>(<span class="kw">assays</span>(sce)<span class="op">$</span>counts[,select])</span>
<span id="cb35-4"><a href="#cb35-4"></a>cs[<span class="kw">order</span>(cs, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)][<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">10</span>)]</span></code></pre></div>
<p>Let’s take a look at how comparable two cells (replicates) of cluster 1 are. We will compare the cell with the highest library size with the cell that has the 10th highest library size using MD-plots (mean-difference plots, as introduced by <a href="https://www.jstor.org/stable/24307038?seq=1#metadata_info_tab_contents">Dudoit et al. (2002)</a>), also sometimes referred to as MA-plots.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>targetCells &lt;-<span class="st"> </span><span class="kw">names</span>(cs[<span class="kw">order</span>(cs, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)][<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">10</span>)])</span>
<span id="cb36-2"><a href="#cb36-2"></a></span>
<span id="cb36-3"><a href="#cb36-3"></a>M &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(<span class="kw">assays</span>(sce)<span class="op">$</span>counts[,targetCells])</span>
<span id="cb36-4"><a href="#cb36-4"></a>D &lt;-<span class="st"> </span><span class="kw">assays</span>(sce)<span class="op">$</span>counts[,targetCells[<span class="dv">2</span>]] <span class="op">/</span><span class="st"> </span><span class="kw">assays</span>(sce)<span class="op">$</span>counts[,targetCells[<span class="dv">1</span>]]</span>
<span id="cb36-5"><a href="#cb36-5"></a><span class="kw">plot</span>(<span class="dt">x =</span> <span class="kw">log</span>(M), <span class="dt">y =</span> <span class="kw">log2</span>(D),</span>
<span id="cb36-6"><a href="#cb36-6"></a>     <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex=</span><span class="dv">1</span><span class="op">/</span><span class="dv">3</span>,</span>
<span id="cb36-7"><a href="#cb36-7"></a>     <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Cell &quot;</span>, targetCells[<span class="dv">2</span>], <span class="st">&quot; vs cell &quot;</span>, targetCells[<span class="dv">1</span>]),</span>
<span id="cb36-8"><a href="#cb36-8"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Log mean&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Log2 fold-change&quot;</span>,</span>
<span id="cb36-9"><a href="#cb36-9"></a>     <span class="dt">bty =</span> <span class="st">&#39;l&#39;</span>)</span>
<span id="cb36-10"><a href="#cb36-10"></a><span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>, <span class="dt">col=</span><span class="st">&quot;orange&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</span></code></pre></div>
<p>We see clear bias in the comparison of the 1st and 10th most deeply sequenced cell from cell cluster 1. We see that the log fold-changes are biased downwards. This means that, on average, a gene is higher expressed in cell 1 versus cell 10. Looking at the library sizes, we can indeed see that the library size for cell 1 is 20869 counts, while it is only 14719 counts for cell 10! This is a clear library size effect that we should take into account.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a><span class="co"># normalize the count data using the previously computed &quot;size factors&quot;</span></span>
<span id="cb37-2"><a href="#cb37-2"></a><span class="kw">assay</span>(sce, <span class="st">&quot;normed&quot;</span>) &lt;-<span class="st"> </span><span class="kw">normalizeCounts</span>(sce, </span>
<span id="cb37-3"><a href="#cb37-3"></a>                                        <span class="dt">log=</span><span class="ot">FALSE</span>,</span>
<span id="cb37-4"><a href="#cb37-4"></a>                                        <span class="dt">size.factors=</span>sf, </span>
<span id="cb37-5"><a href="#cb37-5"></a>                                        <span class="dt">pseudo.count=</span><span class="dv">0</span>)</span>
<span id="cb37-6"><a href="#cb37-6"></a></span>
<span id="cb37-7"><a href="#cb37-7"></a>M &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(<span class="kw">assays</span>(sce)<span class="op">$</span>normed[,targetCells])</span>
<span id="cb37-8"><a href="#cb37-8"></a>D &lt;-<span class="st"> </span><span class="kw">assays</span>(sce)<span class="op">$</span>normed[,targetCells[<span class="dv">2</span>]] <span class="op">/</span><span class="st"> </span><span class="kw">assays</span>(sce)<span class="op">$</span>normed[,targetCells[<span class="dv">1</span>]]</span>
<span id="cb37-9"><a href="#cb37-9"></a><span class="kw">plot</span>(<span class="dt">x =</span> <span class="kw">log</span>(M), <span class="dt">y =</span> <span class="kw">log2</span>(D),</span>
<span id="cb37-10"><a href="#cb37-10"></a>     <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex=</span><span class="dv">1</span><span class="op">/</span><span class="dv">3</span>,</span>
<span id="cb37-11"><a href="#cb37-11"></a>     <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Cell &quot;</span>, targetCells[<span class="dv">2</span>], <span class="st">&quot; vs cell &quot;</span>, targetCells[<span class="dv">1</span>]),</span>
<span id="cb37-12"><a href="#cb37-12"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Log mean&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Log2 fold-change&quot;</span>,</span>
<span id="cb37-13"><a href="#cb37-13"></a>     <span class="dt">bty =</span> <span class="st">&#39;l&#39;</span>)</span>
<span id="cb37-14"><a href="#cb37-14"></a><span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>, <span class="dt">col=</span><span class="st">&quot;orange&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</span></code></pre></div>
<p>Upon normalizing the data using size factors, we have removed bias as a consequence of differences in sequencing depth.</p>
<p>Note that we computed the normalized count matrix only for demonstrating the effect of such normalization. In a typical workflow, the size factors are used as offsets in the downstream models rather than to perform data transformation. In brief, such transformation would distort the mean-variance relationship in the data. For more details, we refer to the document that was touched upon in the first theory session (<a href="https://statomics.github.io/SGA21/sequencing_scalingNormalization.html">link</a>)</p>
</div>
<div id="feature-selection" class="section level1">
<h1><span class="header-section-number">11</span> Feature selection</h1>
<p>As dimensions increase, shortest and farthest distances between points become nearly inseparable. In high-dimensional space, it is therefore extremely difficult to separate signal from noise.</p>
<p>In order to recover structure (e.g. setting up a dimension-reduced space to help us find cell-type clusters in the data), we want to move to an informative, lower-dimensional space. We will select genes which we hope are informative for recovering the biological structure. But what defines an informative gene?</p>
<div id="selecting-genes-with-high-variance" class="section level2">
<h2><span class="header-section-number">11.1</span> Selecting genes with high variance</h2>
<p>The simplest approach to quantifying per-gene variation is to compute the variance of the log-normalized expression values (i.e., “log-counts”) for each gene across all cells.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="co"># calculate variance of log-normalized counts for each gene</span></span>
<span id="cb38-2"><a href="#cb38-2"></a>geneVars &lt;-<span class="st"> </span>genefilter<span class="op">::</span><span class="kw">rowVars</span>(...)</span>
<span id="cb38-3"><a href="#cb38-3"></a><span class="co"># select top 1000 highly variable genes</span></span>
<span id="cb38-4"><a href="#cb38-4"></a>highVarGenes &lt;-<span class="st"> </span><span class="kw">names</span>(geneVars)[<span class="kw">order</span>(geneVars, <span class="dt">decreasing=</span><span class="ot">TRUE</span>)[<span class="dv">1</span><span class="op">:</span><span class="fl">1e3</span>]]</span>
<span id="cb38-5"><a href="#cb38-5"></a><span class="kw">head</span>(highVarGenes)</span></code></pre></div>
</div>
<div id="selecting-genes-with-high-variation-with-respect-to-mean" class="section level2">
<h2><span class="header-section-number">11.2</span> Selecting genes with high variation with respect to mean</h2>
<p>While calculation of the per-gene variance is simple, feature selection requires modelling of the mean-variance relationship. Indeed, not accounting for the mean-variance structure while selecting highly variable genes will oftentimes boil down to selecting the most highly expressed genes. The log-transformation is a helpful variance stabilizing transformation, however, it is not perfect, meaning that the variance of a gene is not completely independent of its mean. Therefore, feature selection may still be driven by average expression rather than underlying biological heterogeneity.</p>
<p>A (rightly so) popular approach is to select genes that have a high variance with respect to their mean. Often, first an empirical mean-variance trend is fitted, upon which genes with the highest positive residuals are selected. Being intuitive, reasonable and fairly straight-forward, this method is widely used.</p>
<p>To account for the mean-variance effect, we use the <code>modelGeneVar</code> function of the <code>scran</code> package to fit a trend to the variance with respect to abundance across all genes (on log-normalized expression values of the sce object).</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="kw">library</span>(scran)</span>
<span id="cb39-2"><a href="#cb39-2"></a><span class="co"># `modelGeneVar` function description: Model the variance of the log-expression </span></span>
<span id="cb39-3"><a href="#cb39-3"></a><span class="co"># profiles for each gene, decomposing it into technical and biological </span></span>
<span id="cb39-4"><a href="#cb39-4"></a><span class="co"># components based on a fitted mean-variance trend.</span></span>
<span id="cb39-5"><a href="#cb39-5"></a>dec &lt;-<span class="st"> </span><span class="kw">modelGeneVar</span>(...) <span class="co"># input a sce object with precomputed logcounts assay</span></span>
<span id="cb39-6"><a href="#cb39-6"></a><span class="kw">head</span>(dec)</span></code></pre></div>
<p>The fitted value for each gene is used as a proxy for the technical component of variation for each gene, under the assumption that most genes exhibit a low baseline level of variation that is not biologically interesting. The biological component of variation for each gene is defined as the the residual from the trend. Ranking genes by the biological component enables identification of interesting genes for downstream analyses in a manner that accounts for the mean-variance relationship.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a>fitRetina &lt;-<span class="st"> </span><span class="kw">metadata</span>(...)</span>
<span id="cb40-2"><a href="#cb40-2"></a><span class="kw">plot</span>(<span class="dt">x =</span> ..., <span class="co"># the fitted means on the x-axis </span></span>
<span id="cb40-3"><a href="#cb40-3"></a>     <span class="dt">y =</span> ..., <span class="co"># the fitted variances on the y-axis</span></span>
<span id="cb40-4"><a href="#cb40-4"></a>     <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>,</span>
<span id="cb40-5"><a href="#cb40-5"></a>    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</span>
<span id="cb40-6"><a href="#cb40-6"></a><span class="kw">curve</span>(fitRetina<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="dv">2</span>) <span class="co"># adds a trend line</span></span></code></pre></div>
<p>We are interested in those genes for which the variance in expression is higher than what we would expect for that gene based on its mean expression.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a><span class="co"># get 10% most variable genes</span></span>
<span id="cb41-2"><a href="#cb41-2"></a><span class="co"># Based on the output statistics of modelGeneVar, select the 10% most variable genes</span></span>
<span id="cb41-3"><a href="#cb41-3"></a>hvg &lt;-<span class="st"> </span><span class="kw">getTopHVGs</span>(<span class="dt">stats =</span> ..., </span>
<span id="cb41-4"><a href="#cb41-4"></a>                  <span class="dt">prop =</span> ...) <span class="co"># percentage of top variable genes. Alternatively,</span></span>
<span id="cb41-5"><a href="#cb41-5"></a><span class="co"># n = ... could be used to select the top number of variable genes</span></span>
<span id="cb41-6"><a href="#cb41-6"></a><span class="kw">head</span>(hvg)</span>
<span id="cb41-7"><a href="#cb41-7"></a></span>
<span id="cb41-8"><a href="#cb41-8"></a><span class="co"># plot these </span></span>
<span id="cb41-9"><a href="#cb41-9"></a><span class="kw">plot</span>(<span class="dt">x =</span> ..., <span class="co"># the fitted means on the x-axis </span></span>
<span id="cb41-10"><a href="#cb41-10"></a>     <span class="dt">y =</span> ..., <span class="co"># the fitted variances on the y-axis</span></span>
<span id="cb41-11"><a href="#cb41-11"></a>     <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;orange&quot;</span>, <span class="st">&quot;darkseagreen3&quot;</span>)[(<span class="kw">names</span>(fitRetina<span class="op">$</span>mean) <span class="op">%in%</span><span class="st"> </span>hvg)<span class="op">+</span><span class="dv">1</span>],</span>
<span id="cb41-12"><a href="#cb41-12"></a>     <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>,</span>
<span id="cb41-13"><a href="#cb41-13"></a>    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</span>
<span id="cb41-14"><a href="#cb41-14"></a><span class="kw">curve</span>(fitRetina<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</span>
<span id="cb41-15"><a href="#cb41-15"></a><span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>, </span>
<span id="cb41-16"><a href="#cb41-16"></a>       <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&quot;Selected&quot;</span>, <span class="st">&quot;Not selected&quot;</span>), </span>
<span id="cb41-17"><a href="#cb41-17"></a>       <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;darkseagreen3&quot;</span>, <span class="st">&quot;orange&quot;</span>),</span>
<span id="cb41-18"><a href="#cb41-18"></a>       <span class="dt">pch =</span> <span class="dv">16</span>,</span>
<span id="cb41-19"><a href="#cb41-19"></a>       <span class="dt">bty=</span><span class="st">&#39;n&#39;</span>)</span></code></pre></div>
<p>As a comparison, we could color the genes on this figure according to the selection that was made purely by looking at the raw variance of each gene.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a>...</span></code></pre></div>
<p>As expected, we here simply select genes with a high variance, without recognizing that a high variance is typically driven by a high mean.</p>
</div>
<div id="high-deviance-genes" class="section level2">
<h2><span class="header-section-number">11.3</span> High deviance genes</h2>
<p>Here, we will select genes with a high residual deviance. The idea is that we assume a null model of constant expression fraction (i.e., RNA concentration) across all cells. We subsequently calculate a goodness-of-fit statistic for each gene, assessing whether the model is a good approximation to the gene expression of the corresponding gene.</p>
<p>If the model fits poorly, i.e., the gene has a high deviance, the expression fraction varies significantly across the cells in our datasets.Genes with a high deviance will thus most poorly fit a null model where the relative abundance is equal for all cells, which therefore are informative.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a><span class="co">#BiocManager::install(&quot;scry&quot;)</span></span>
<span id="cb43-2"><a href="#cb43-2"></a><span class="kw">library</span>(scry)</span>
<span id="cb43-3"><a href="#cb43-3"></a><span class="co"># `devianceFeatureSelection` function description: Computes a deviance statistic </span></span>
<span id="cb43-4"><a href="#cb43-4"></a><span class="co"># for each row feature (such as a gene) for count data based on a multinomial </span></span>
<span id="cb43-5"><a href="#cb43-5"></a><span class="co"># null model [...]. </span></span>
<span id="cb43-6"><a href="#cb43-6"></a>sce &lt;-<span class="st"> </span><span class="kw">devianceFeatureSelection</span>(<span class="dt">object =</span> ..., </span>
<span id="cb43-7"><a href="#cb43-7"></a>                                <span class="dt">assay =</span> ...) <span class="co"># computed on raw counts</span></span>
<span id="cb43-8"><a href="#cb43-8"></a></span>
<span id="cb43-9"><a href="#cb43-9"></a><span class="kw">plot</span>(<span class="kw">sort</span>(<span class="kw">rowData</span>(sce)<span class="op">$</span>binomial_deviance, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>), </span>
<span id="cb43-10"><a href="#cb43-10"></a>     <span class="dt">type=</span><span class="st">&quot;l&quot;</span>, </span>
<span id="cb43-11"><a href="#cb43-11"></a>     <span class="dt">xlab=</span><span class="st">&quot;ranked genes&quot;</span>, </span>
<span id="cb43-12"><a href="#cb43-12"></a>     <span class="dt">ylab=</span><span class="st">&quot;binomial deviance&quot;</span>, </span>
<span id="cb43-13"><a href="#cb43-13"></a>     <span class="dt">main=</span><span class="st">&quot;Feature Selection with Deviance&quot;</span>)</span>
<span id="cb43-14"><a href="#cb43-14"></a><span class="kw">abline</span>(<span class="dt">v=</span><span class="dv">2000</span>, <span class="dt">lty=</span><span class="dv">2</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p>Our plot looks similar to one displayed in the <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/scry/inst/doc/scry.html">vignette of the scry package</a>. Based on that plot, the authors suggest retaining 2.000 genes (the top 2000 based on the deviance residuals) for downstream dimensionality reduction and clustering.</p>
</div>
<div id="seurat-vst" class="section level2">
<h2><span class="header-section-number">11.4</span> Seurat VST</h2>
<p>Another very common feature selection strategy is the variance-stabilizing transformation from the <code>Seurat</code> R package, which amounts to calculating Pearson residuals from a regularized negative binomial regression model, with sequencing depth as a covariate.</p>
<p><strong>Intermezzo: interoperability between SingleCellExperiment and Seurat objects</strong></p>
<p>In this lecture series, we always make use of the SingleCellExperiment object and the packages available from Bioconductor. Another very popular toolbox for performing scRNA-seq data analysis is <code>Seurat</code>. However, functions from <code>Seurat</code> cannot be used directly to manipulate <code>SingleCellExperiment</code> objects, and vice versa. Fortunately, efforts have been made to increase the interoperability between the two toolboxes.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a><span class="kw">library</span>(Seurat)</span>
<span id="cb44-2"><a href="#cb44-2"></a>seurat_obj &lt;-<span class="st"> </span><span class="kw">as.Seurat</span>(...) <span class="co"># sce object</span></span>
<span id="cb44-3"><a href="#cb44-3"></a>seurat_obj <span class="co"># notice the &quot;0 variable features&quot;</span></span></code></pre></div>
<p>On this object, we may use functions from the <code>Seurat</code> toolbox. For instance, we may search for highly variable features using <code>Seurat</code>’s VST implementation:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a><span class="co"># Prior to detecting variable features using VST, Seurat normalizes the data.</span></span>
<span id="cb45-2"><a href="#cb45-2"></a><span class="co"># The `NormalizeData` function takes as input the Seurat object, and uses </span></span>
<span id="cb45-3"><a href="#cb45-3"></a><span class="co"># log normalization and a scale factor of 10000 as default (see help file).</span></span>
<span id="cb45-4"><a href="#cb45-4"></a>seurat_obj &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">NormalizeData</span>(<span class="dt">object =</span> ..., </span>
<span id="cb45-5"><a href="#cb45-5"></a>                                    <span class="dt">normalization.method =</span> ..., </span>
<span id="cb45-6"><a href="#cb45-6"></a>                                    <span class="dt">scale.factor =</span> ...)</span>
<span id="cb45-7"><a href="#cb45-7"></a></span>
<span id="cb45-8"><a href="#cb45-8"></a><span class="co"># Next, perform feature selection based on VST</span></span>
<span id="cb45-9"><a href="#cb45-9"></a>seurat_obj &lt;-<span class="st">  </span><span class="kw">FindVariableFeatures</span>(<span class="dt">object =</span> ...,</span>
<span id="cb45-10"><a href="#cb45-10"></a>                                    <span class="dt">selection.method =</span> ...)</span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1"></a>seurat_obj  <span class="co"># notice the &quot;2000 variable features&quot; (default)</span></span>
<span id="cb46-2"><a href="#cb46-2"></a><span class="kw">head</span>(<span class="kw">VariableFeatures</span>(seurat_obj)) <span class="co"># here they are</span></span></code></pre></div>
</div>
</div>
<div id="dimensionality-reduction" class="section level1">
<h1><span class="header-section-number">12</span> Dimensionality reduction</h1>
<p>Note that, below, we color the cells using the known, true cell type label as defined in the metadata, to empirically evaluate the dimensionality reduction. In reality, we don’t know this yet at this stage.</p>
<div id="the-most-basic-dr" class="section level2">
<h2><span class="header-section-number">12.1</span> The most basic DR</h2>
<p>Just by looking at the top two genes based on our feature selection criterion, we can already see some separation according to the cell type!</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a><span class="kw">colData</span>(sce)<span class="op">$</span>cluster &lt;-<span class="st"> </span><span class="kw">as.factor</span>(<span class="kw">colData</span>(sce)<span class="op">$</span>cluster)</span>
<span id="cb47-2"><a href="#cb47-2"></a>cl &lt;-<span class="st"> </span><span class="kw">colData</span>(sce)<span class="op">$</span>cluster</span>
<span id="cb47-3"><a href="#cb47-3"></a><span class="kw">par</span>(<span class="dt">bty=</span><span class="st">&#39;l&#39;</span>)</span>
<span id="cb47-4"><a href="#cb47-4"></a><span class="kw">plot</span>(<span class="dt">x =</span> ..., <span class="co"># extract the counts for the most variable gene</span></span>
<span id="cb47-5"><a href="#cb47-5"></a>     <span class="dt">y =</span> ..., <span class="co"># extract the counts for the second most variable gene</span></span>
<span id="cb47-6"><a href="#cb47-6"></a>     <span class="dt">col =</span> <span class="kw">as.numeric</span>(cl),</span>
<span id="cb47-7"><a href="#cb47-7"></a>     <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">3</span>,</span>
<span id="cb47-8"><a href="#cb47-8"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Most informative gene&quot;</span>,</span>
<span id="cb47-9"><a href="#cb47-9"></a>     <span class="dt">ylab =</span> <span class="st">&quot;Second most informative gene&quot;</span>,</span>
<span id="cb47-10"><a href="#cb47-10"></a>     <span class="dt">main =</span> <span class="st">&quot;Cells colored acc to cell type&quot;</span>)</span></code></pre></div>
<p>We are able to recover quite some structure. However, many cell populations remain obscure, and the plot is overcrowded.</p>
</div>
<div id="linear-dimensionality-reduction-pca" class="section level2">
<h2><span class="header-section-number">12.2</span> Linear dimensionality reduction: PCA</h2>
<p>A DR method is linear when the reduced dimensions are a linear function of the original variables. For example, in PCA, each principal component is a linear combination of genes, therefore the DR is a linear function of the original variables.</p>
<p>Typically, PCA is performed on log-transformed normalized counts. The log-transformation helps somewhat, but not completely, to account for the mean-variance relationship. PCA works well for bulk RNA-seq data. However, the structure of scRNA-seq data is often too complex to be visualized by a small number of PCs.</p>
<p>There are several R functions that allow you to perform PCA. Here, we make use of the <code>runPCA</code> function of the <code>scater</code> package, which has been specifically developed for performing PCA on <code>SingleCellExperiment</code> objects. We calculate the top 30 principal components.</p>
<div id="pca-with-feature-selection" class="section level3">
<h3><span class="header-section-number">12.2.1</span> PCA with feature selection</h3>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1"></a><span class="co"># Perform a PCA for our data. Compute the first 30 principal components, only</span></span>
<span id="cb48-2"><a href="#cb48-2"></a><span class="co"># using the selected features with high variance with respect to the mean.</span></span>
<span id="cb48-3"><a href="#cb48-3"></a><span class="kw">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb48-4"><a href="#cb48-4"></a>sce &lt;-<span class="st"> </span><span class="kw">runPCA</span>(...) <span class="co"># you need three arguments</span></span></code></pre></div>
<p>PCA has been performed. The PCA information has been automatically stored in the <em>reducedDim</em> slot of the SingleCellExperiment object.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a><span class="kw">reducedDimNames</span>(sce)</span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1"></a><span class="kw">head</span>(<span class="kw">reducedDim</span>(sce,</span>
<span id="cb50-2"><a href="#cb50-2"></a>                <span class="dt">type=</span><span class="st">&quot;PCA&quot;</span>))</span></code></pre></div>
<p>The <code>plotPCA</code> function of the <code>scater</code> package now allows us to visualize the cells in PCA space, based on the PCA information stored in our object:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a><span class="kw">plotPCA</span>(sce, </span>
<span id="cb51-2"><a href="#cb51-2"></a>        <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>)</span></code></pre></div>
<p>While the large number of clusters in this dataset makes it hard to distinguish between all the different colors, we can already see that PCA retrieves some, but not all, of the structure in the data that was discovered by the original authors.</p>
<p>How many of the top PCs should we retain for downstream analyses? The choice of the number of PCs is a decision that is analogous to the choice of the number of HVGs to use. Using more PCs will retain more biological signal at the cost of including more noise that might mask said signal. On the other hand, using fewer PCs will introduce competition between different factors of variation, where weaker (but still interesting) factors may be pushed down into lower PCs and inadvertently discarded from downstream analyses.</p>
<p>Most analysts will simply aim to use a “reasonable” but arbitrary value, typically ranging from 10 to 50. This is often satisfactory as the later PCs explain so little variance that their inclusion or omission has no major effect.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1"></a>percent.var &lt;-<span class="st"> </span><span class="kw">attr</span>(<span class="kw">reducedDim</span>(sce), <span class="st">&quot;percentVar&quot;</span>)</span>
<span id="cb52-2"><a href="#cb52-2"></a><span class="kw">plot</span>(percent.var, <span class="dt">log=</span><span class="st">&quot;y&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;PC&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Variance explained (%)&quot;</span>)</span>
<span id="cb52-3"><a href="#cb52-3"></a><span class="kw">plot</span>(<span class="kw">cumsum</span>(percent.var), <span class="dt">xlab=</span><span class="st">&quot;PC&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Cumulative variance explained (%)&quot;</span>)</span></code></pre></div>
<p>Here, retaining ±15PCs seems reasonable. If you really prefer a more data-driven way for determining this, there are procedures available that aim to computationally identify the elbow/knee point in the variance explained per PC plot.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1"></a><span class="kw">library</span>(PCAtools)</span>
<span id="cb53-2"><a href="#cb53-2"></a>chosen.elbow &lt;-<span class="st"> </span><span class="kw">findElbowPoint</span>(percent.var)</span>
<span id="cb53-3"><a href="#cb53-3"></a>chosen.elbow</span></code></pre></div>
</div>
<div id="pca-without-feature-selection" class="section level3">
<h3><span class="header-section-number">12.2.2</span> PCA without feature selection</h3>
<p>Note: more features -&gt; computationally more intensive!</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1"></a><span class="kw">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb54-2"><a href="#cb54-2"></a>sceNoFS &lt;-<span class="st"> </span><span class="kw">runPCA</span>(<span class="dt">x =</span> ..., </span>
<span id="cb54-3"><a href="#cb54-3"></a>                  <span class="dt">ncomponents =</span> ..., </span>
<span id="cb54-4"><a href="#cb54-4"></a>                  <span class="dt">subset_row =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sce))</span>
<span id="cb54-5"><a href="#cb54-5"></a><span class="kw">plotPCA</span>(...)</span>
<span id="cb54-6"><a href="#cb54-6"></a><span class="kw">rm</span>(sceNoFS) <span class="co"># remove the object, we don&#39;t need it anymore</span></span></code></pre></div>
<p>While we use more information to make this PCA plot (17.887 genes) as compared to the feature selected PCA plot (642 genes), we seem to retrieve less structure in the data. This is the power of feature selection, an increase in the signal-to-noise ratio!</p>
</div>
<div id="effect-of-feature-selection-on-pca" class="section level3">
<h3><span class="header-section-number">12.2.3</span> Effect of feature selection on PCA</h3>
<p>First, we compare the different feature selection criteria, using the top 1000 highly variable genes for each method.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1"></a><span class="co"># Get top 1000 highly variable features using the highly variable genes</span></span>
<span id="cb55-2"><a href="#cb55-2"></a><span class="co"># (w.r.t. mean), high deviance genes and the VST strategy of Seurat</span></span>
<span id="cb55-3"><a href="#cb55-3"></a></span>
<span id="cb55-4"><a href="#cb55-4"></a>hvg1000 &lt;-<span class="st"> </span>... <span class="co"># extract top 1000 highly variable genes</span></span>
<span id="cb55-5"><a href="#cb55-5"></a>hdg1000 &lt;-<span class="st"> </span>... <span class="co"># extract top 1000 genes with high deviance</span></span>
<span id="cb55-6"><a href="#cb55-6"></a>vst1000 &lt;-<span class="st"> </span>... <span class="co"># extract top 1000 genes based on VST analysis</span></span>
<span id="cb55-7"><a href="#cb55-7"></a></span>
<span id="cb55-8"><a href="#cb55-8"></a><span class="co"># HVG strategy</span></span>
<span id="cb55-9"><a href="#cb55-9"></a><span class="kw">plotPCA</span>(</span>
<span id="cb55-10"><a href="#cb55-10"></a>    <span class="kw">runPCA</span>(sce,</span>
<span id="cb55-11"><a href="#cb55-11"></a>           <span class="dt">ncomponents =</span> <span class="dv">2</span>, </span>
<span id="cb55-12"><a href="#cb55-12"></a>           <span class="dt">subset_row =</span> hvg1000), </span>
<span id="cb55-13"><a href="#cb55-13"></a>    <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>) <span class="op">+</span></span>
<span id="cb55-14"><a href="#cb55-14"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;Highly variable genes&quot;</span>)</span>
<span id="cb55-15"><a href="#cb55-15"></a></span>
<span id="cb55-16"><a href="#cb55-16"></a><span class="co"># HDG strategy</span></span>
<span id="cb55-17"><a href="#cb55-17"></a><span class="kw">plotPCA</span>(</span>
<span id="cb55-18"><a href="#cb55-18"></a>    <span class="kw">runPCA</span>(sce,</span>
<span id="cb55-19"><a href="#cb55-19"></a>           <span class="dt">ncomponents =</span> <span class="dv">2</span>, </span>
<span id="cb55-20"><a href="#cb55-20"></a>           <span class="dt">subset_row =</span> hdg1000), </span>
<span id="cb55-21"><a href="#cb55-21"></a>    <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>) <span class="op">+</span></span>
<span id="cb55-22"><a href="#cb55-22"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;High-deviance genes&quot;</span>)</span>
<span id="cb55-23"><a href="#cb55-23"></a></span>
<span id="cb55-24"><a href="#cb55-24"></a><span class="co"># VST strategy</span></span>
<span id="cb55-25"><a href="#cb55-25"></a><span class="kw">plotPCA</span>(</span>
<span id="cb55-26"><a href="#cb55-26"></a>    <span class="kw">runPCA</span>(sce,</span>
<span id="cb55-27"><a href="#cb55-27"></a>           <span class="dt">ncomponents =</span> <span class="dv">2</span>, </span>
<span id="cb55-28"><a href="#cb55-28"></a>           <span class="dt">subset_row =</span> vst1000), </span>
<span id="cb55-29"><a href="#cb55-29"></a>    <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>,</span>
<span id="cb55-30"><a href="#cb55-30"></a>    ) <span class="op">+</span></span>
<span id="cb55-31"><a href="#cb55-31"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;Seurat VST&quot;</span>)</span></code></pre></div>
<p>Next, we assess the sensitivity on the number of top features for the highly variable genes method;</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1"></a>hvg_all &lt;-<span class="st"> </span><span class="kw">getTopHVGs</span>(dec)</span>
<span id="cb56-2"><a href="#cb56-2"></a></span>
<span id="cb56-3"><a href="#cb56-3"></a>hvg2000 &lt;-<span class="st"> </span>... <span class="co"># extract top 2000 highly variable genes</span></span>
<span id="cb56-4"><a href="#cb56-4"></a>hvg1000 &lt;-<span class="st"> </span>... <span class="co"># extract top 1000 highly variable genes</span></span>
<span id="cb56-5"><a href="#cb56-5"></a>hvg100 &lt;-<span class="st"> </span>... <span class="co"># extract top 100 highly variable genes</span></span>
<span id="cb56-6"><a href="#cb56-6"></a>hvg10 &lt;-<span class="st"> </span>... <span class="co"># extract top 10 highly variable genes</span></span>
<span id="cb56-7"><a href="#cb56-7"></a>hvg5 &lt;-<span class="st"> </span>... <span class="co"># extract top 5 highly variable genes</span></span>
<span id="cb56-8"><a href="#cb56-8"></a></span>
<span id="cb56-9"><a href="#cb56-9"></a><span class="kw">plotPCA</span>(</span>
<span id="cb56-10"><a href="#cb56-10"></a>    <span class="kw">runPCA</span>(sce,</span>
<span id="cb56-11"><a href="#cb56-11"></a>           <span class="dt">ncomponents =</span> <span class="dv">2</span>), </span>
<span id="cb56-12"><a href="#cb56-12"></a>    <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>) <span class="op">+</span></span>
<span id="cb56-13"><a href="#cb56-13"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;All genes&quot;</span>)</span>
<span id="cb56-14"><a href="#cb56-14"></a></span>
<span id="cb56-15"><a href="#cb56-15"></a><span class="kw">plotPCA</span>(</span>
<span id="cb56-16"><a href="#cb56-16"></a>    <span class="kw">runPCA</span>(sce,</span>
<span id="cb56-17"><a href="#cb56-17"></a>           <span class="dt">ncomponents =</span> <span class="dv">2</span>, </span>
<span id="cb56-18"><a href="#cb56-18"></a>           <span class="dt">subset_row =</span> hvg2000), </span>
<span id="cb56-19"><a href="#cb56-19"></a>    <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>) <span class="op">+</span></span>
<span id="cb56-20"><a href="#cb56-20"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;Top 2000 genes&quot;</span>)</span>
<span id="cb56-21"><a href="#cb56-21"></a></span>
<span id="cb56-22"><a href="#cb56-22"></a><span class="kw">plotPCA</span>(</span>
<span id="cb56-23"><a href="#cb56-23"></a>    <span class="kw">runPCA</span>(sce,</span>
<span id="cb56-24"><a href="#cb56-24"></a>           <span class="dt">ncomponents =</span> <span class="dv">2</span>, </span>
<span id="cb56-25"><a href="#cb56-25"></a>           <span class="dt">subset_row =</span> hvg1000), </span>
<span id="cb56-26"><a href="#cb56-26"></a>    <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>) <span class="op">+</span></span>
<span id="cb56-27"><a href="#cb56-27"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;Top 1000 genes&quot;</span>)</span>
<span id="cb56-28"><a href="#cb56-28"></a></span>
<span id="cb56-29"><a href="#cb56-29"></a><span class="kw">plotPCA</span>(</span>
<span id="cb56-30"><a href="#cb56-30"></a>    <span class="kw">runPCA</span>(sce,</span>
<span id="cb56-31"><a href="#cb56-31"></a>           <span class="dt">ncomponents =</span> <span class="dv">2</span>, </span>
<span id="cb56-32"><a href="#cb56-32"></a>           <span class="dt">subset_row =</span> hvg100), </span>
<span id="cb56-33"><a href="#cb56-33"></a>    <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>) <span class="op">+</span></span>
<span id="cb56-34"><a href="#cb56-34"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;Top 100 genes&quot;</span>)</span>
<span id="cb56-35"><a href="#cb56-35"></a></span>
<span id="cb56-36"><a href="#cb56-36"></a><span class="kw">plotPCA</span>(</span>
<span id="cb56-37"><a href="#cb56-37"></a>    <span class="kw">runPCA</span>(sce,</span>
<span id="cb56-38"><a href="#cb56-38"></a>           <span class="dt">ncomponents =</span> <span class="dv">2</span>, </span>
<span id="cb56-39"><a href="#cb56-39"></a>           <span class="dt">subset_row =</span> hvg10), </span>
<span id="cb56-40"><a href="#cb56-40"></a>    <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>) <span class="op">+</span></span>
<span id="cb56-41"><a href="#cb56-41"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;Top 10 genes&quot;</span>)</span>
<span id="cb56-42"><a href="#cb56-42"></a></span>
<span id="cb56-43"><a href="#cb56-43"></a><span class="kw">plotPCA</span>(</span>
<span id="cb56-44"><a href="#cb56-44"></a>    <span class="kw">runPCA</span>(sce,</span>
<span id="cb56-45"><a href="#cb56-45"></a>           <span class="dt">ncomponents =</span> <span class="dv">2</span>, </span>
<span id="cb56-46"><a href="#cb56-46"></a>           <span class="dt">subset_row =</span> hvg5), </span>
<span id="cb56-47"><a href="#cb56-47"></a>    <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>) <span class="op">+</span></span>
<span id="cb56-48"><a href="#cb56-48"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;Top 5 genes&quot;</span>)</span></code></pre></div>
</div>
<div id="remarks-on-pca" class="section level3">
<h3><span class="header-section-number">12.2.4</span> Remarks on PCA</h3>
<p>Visualizations of reduced dimensions from linear dimensionality reduction methods are often “overcrowded”, and it is hard to see structure (e.g., the PCA plot we just saw). Non-linear dimensionality reduction methods can overcome this problem. As the name suggests, the reduced dimensions are a non-linear function of the observed data. We will not go into detail as to how these work under the hood, but provide a few guidelines for the most popular methods. Often, the top (~10-50) PCs are provided as input.</p>
</div>
</div>
<div id="a-generalization-of-pca-for-exponential-family-distributions." class="section level2">
<h2><span class="header-section-number">12.3</span> A generalization of PCA for exponential family distributions.</h2>
<p>PCA is implicitly based on Euclidean distances, corresponding to maximizing a Gaussian likelihood, which is inappropriate for count data such as scRNA-seq. <a href="https://doi.org/10.1186/s13059-019-1861-6">Townes et al.</a> (2019) develop GLM-PCA, a generalization of PCA for exponential family likelihoods. They posit, using negative control data, that the data generative mechanism of UMI count data can be considered to be multinomial.</p>
<p>The GLM-PCA strategy is implemented in the <code>glmpca</code> function of the <code>glmpca</code> package.</p>
<p>Note that this function is quite computationally intensive. For regular PCA on log-transformed normalized counts, the underlying computations can be strongly simplified. Here, we work with the raw counts, which we assume to be <em>Poisson</em> distributed, requiring an iterative optimization scheme.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1"></a><span class="co"># runs 2min on my laptop</span></span>
<span id="cb57-2"><a href="#cb57-2"></a><span class="kw">library</span>(glmpca)</span>
<span id="cb57-3"><a href="#cb57-3"></a><span class="kw">set.seed</span>(<span class="dv">211103</span>)</span>
<span id="cb57-4"><a href="#cb57-4"></a>poipca &lt;-<span class="st"> </span><span class="kw">glmpca</span>(<span class="dt">Y =</span> ..., <span class="co"># counts for the higly variable genes</span></span>
<span id="cb57-5"><a href="#cb57-5"></a>                 <span class="dt">L =</span> ..., <span class="co"># two latent dimensions (components)</span></span>
<span id="cb57-6"><a href="#cb57-6"></a>                 <span class="dt">fam =</span> ..., <span class="co"># assume Poisson for droplet data</span></span>
<span id="cb57-7"><a href="#cb57-7"></a>                 <span class="dt">minibatch =</span> <span class="st">&quot;stochastic&quot;</span>) <span class="co"># used in the package vignette by </span></span>
<span id="cb57-8"><a href="#cb57-8"></a><span class="co"># the authors, but beyond the scope of this session.</span></span>
<span id="cb57-9"><a href="#cb57-9"></a><span class="kw">reducedDim</span>(sce, <span class="st">&quot;PoiPCA&quot;</span>) &lt;-<span class="st"> </span>poipca<span class="op">$</span>factors</span>
<span id="cb57-10"><a href="#cb57-10"></a><span class="kw">plotReducedDim</span>(...) <span class="co"># make sure to specify the right value for the `dimred` argument</span></span></code></pre></div>
<p>Alternatively, we could adopt the GLM-PCA strategy using only the genes with high deviance.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1"></a><span class="co"># Based on the diagnostic plot from the feature selection, we would select the </span></span>
<span id="cb58-2"><a href="#cb58-2"></a><span class="co"># top 2000 genes with highest deviance. However, to reduce the running time of</span></span>
<span id="cb58-3"><a href="#cb58-3"></a><span class="co"># the function, I here select the top 500 (which still is quite slow - 5min).</span></span>
<span id="cb58-4"><a href="#cb58-4"></a>hdg500 &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">sort</span>(<span class="kw">rowData</span>(sce)<span class="op">$</span>binomial_deviance, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>))[<span class="dv">1</span><span class="op">:</span><span class="dv">500</span>]</span>
<span id="cb58-5"><a href="#cb58-5"></a></span>
<span id="cb58-6"><a href="#cb58-6"></a>...</span></code></pre></div>
<p>The authors of the <code>glmpca</code> package note that GLM-PCA can be slow for large datasets. Therefore, they have implemented a fast approximation of the algorithm, which first fits a null model of constant expression for each gene across all cells, and subsequently fits standard PCA to either the Pearson or deviance residuals from the null model.</p>
<p>However, at least for me the <code>nullResiduals</code> function was <strong>extremely slow</strong> even slower than the <code>glmpca</code> code above. Therefore, <strong>I suggest not running</strong> this code, but I leave it in for reference.</p>
<pre><code>sce &lt;- nullResiduals(sce, assay=&quot;counts&quot;, type=&quot;deviance&quot;)
sce &lt;- nullResiduals(sce, assay=&quot;counts&quot;, type=&quot;pearson&quot;)

pca&lt;-function(Y, L=2, center=TRUE, scale=TRUE){
    #assumes features=rows, observations=cols
    res&lt;-prcomp(as.matrix(t(Y)), center=center, scale.=scale, rank.=L)
    factors&lt;-as.data.frame(res$x)
    colnames(factors)&lt;-paste0(&quot;dim&quot;, 1:L)
    factors
}
pca_d &lt;- pca(assay(sce[hdg,], &quot;binomial_deviance_residuals&quot;))
pca_d$resid_type &lt;- &quot;deviance_residuals&quot;
pca_p &lt;- pca(assay(sce[hdg,], &quot;binomial_pearson_residuals&quot;))
pca_p$resid_type &lt;- &quot;pearson_residuals&quot;
cm &lt;- as.data.frame(colData(sce[hdg,]))
pd &lt;- rbind(cbind(cm, pca_d), cbind(cm, pca_p))
ggplot(pd, aes(x=dim1, y=dim2, colour=phenoid)) + geom_point() +
  facet_wrap(~resid_type, scales=&quot;free&quot;, nrow=2) +
  ggtitle(&quot;PCA applied to null residuals of high deviance genes&quot;)</code></pre>
</div>
<div id="non-linear-dimensionality-reduction-t-sne" class="section level2">
<h2><span class="header-section-number">12.4</span> Non-linear dimensionality reduction: t-SNE</h2>
<p>t-SNE focuses on preserving local rather than global distances. Therefore, distances on a t-SNE reduced dimension plot can only be interpreted locally, i.e., cells that are close together in reduced dimension will have a similar transcriptome, but cells that are far away may not necessarily have a very distinct transcriptome.</p>
<p>Running t-SNE on a <code>SingleCellExperiment</code> object can be achieved with the <code>runTSNE</code> function of the <code>scater</code> package. By default, this function will first perform PCA, and use the top 50 PCs as an input to the actual t-SNE algorithm. Since we already performed PCA, we may set <code>dimred = "PCA"</code> as argument to the function. As such we will be performing a T-SNE on the 30 PCs we computed before. If we would like to run t-SNE only using the top 10 PCs, we could set <code>n_dimred = 10</code>.</p>
<p>In addition, we may wish to set <code>external_neighbors=TRUE</code>, which increases the speed of the algorithm for large datasets by applying a heuristic.</p>
<p><strong>Note:</strong> for me this was the slowest function of the analysis (so far). If you feel like your PC/laptop had a lot of trouble with the previous step(s), you may consider not running this code, or running it on a subset of the data (e.g., randomly subsampling cells) for demonstrational purposes. Alternatively, you may consider reducing the input space for the t-SNE algorithm, e.g. by setting <code>n_dimred = 5</code>.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1"></a><span class="co"># (For me it takes 3min30 with n_dimred = 5)</span></span>
<span id="cb60-2"><a href="#cb60-2"></a>sce &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(<span class="dt">x =</span> ..., </span>
<span id="cb60-3"><a href="#cb60-3"></a>               <span class="dt">dimred =</span> ...,</span>
<span id="cb60-4"><a href="#cb60-4"></a>               <span class="dt">n_dimred =</span> ...,</span>
<span id="cb60-5"><a href="#cb60-5"></a>               <span class="dt">external_neighbors =</span> ...)</span>
<span id="cb60-6"><a href="#cb60-6"></a><span class="kw">plotTSNE</span>(sce,</span>
<span id="cb60-7"><a href="#cb60-7"></a>         <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>)</span></code></pre></div>
</div>
<div id="non-linear-dimensionality-reduction-umap" class="section level2">
<h2><span class="header-section-number">12.5</span> Non-linear dimensionality reduction: UMAP</h2>
<p>It is often suggested that UMAP is better than t-SNE in preserving global differences. Therefore, UMAP is also often used in analyses such as trajectory inference, where this is important.</p>
<p>Running UMAP on a <code>SingleCellExperiment</code> object can be achieved with the <code>runUMAP</code> function of the <code>scater</code> package.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1"></a><span class="co"># Using top 10% highly variable genes and top 30 PCs</span></span>
<span id="cb61-2"><a href="#cb61-2"></a>sce &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(...)</span>
<span id="cb61-3"><a href="#cb61-3"></a><span class="kw">plotUMAP</span>(...)</span>
<span id="cb61-4"><a href="#cb61-4"></a></span>
<span id="cb61-5"><a href="#cb61-5"></a><span class="co"># Using top 10% highly variable genes and top 10 PCs</span></span>
<span id="cb61-6"><a href="#cb61-6"></a><span class="kw">plotUMAP</span>(<span class="kw">runUMAP</span>(...),</span>
<span id="cb61-7"><a href="#cb61-7"></a>  <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>)</span></code></pre></div>
</div>
</div>

<div id="rmd-source-code">---
title: 'Normalization, feature selection and dimension reduction for the Macosko dataset'
author: "Koen Van den Berge and Jeroen Gilis"
date: "30/11/2021"
output: 
  html_document:
    code_download: true
    toc: true
    toc_float: true
---

# Preamble: installation of Bioconductor libraries

```{r}
# install BiocManager package if not installed yet.
# BiocManager is the package installer for Bioconductor software.
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# install packages if not yet installed.
pkgs <- c("SingleCellExperiment",
          "ExperimentHub",
          "edgeR",
          "biomaRt",
          "DropletUtils", 
          "scRNAseq", 
          "scater", 
          "scuttle", 
          "scran",
          "scry",
          "BiocSingular", 
          "scDblFinder",
          "glmpca",
          "PCAtools",
          "Seurat")
notInstalled <- pkgs[!pkgs %in% installed.packages()[,1]]
if(length(notInstalled) > 0){
  BiocManager::install(notInstalled)
}

```

# The Macosko dataset

In this workshop session, we will analyze the single-cell RNA-seq dataset
from the publication by Macosko *et al.*, Cell 161, 1202–1214 from 2015
[(link)](https://doi.org/10.1016/j.cell.2015.05.002). This is the manuscript in
which the droplet scRNA-seq technology **Drop-seq** was introduced.
Six years after the original publication, drop-seq is still one of the most 
commonly adopted scRNA-seq protocols, as evidenced by the
large number of citations for Macosko *et al.* 
(4.303 citations at November 3, 2021).

The basic idea behind the Drop-seq protocol can be taken from the graphical
abstract of the publication.

```
knitr::include_graphics("macosko_graphicalAbstract.jpeg")
```

The success of Drop-seq can be explained by the following advantageous 
features:

- The use of unique molecular identifiers (UMIs). By working with UMIs, one count
corresponds to one observed mRNA molecule present in the cell. Thanks to the use
of UMI barcodes, PCR artifacts are reduced.

- Scalability: microfluidics technology allows for performing the library prep
reactions inside nanodroplets, in which single cells may be contained. Library prep occurs across all droplets simultaneously.

- Cost: the experiment costs around 6.5 cents (USD) per cell.

- Speed: The very large dataset that we will be working with today was 
generated in an experiment that took only 4 days.

In this particular experiment, Macosko *et al.* sequenced 49,300 cells from the
mouse retina, identifying 39 transcriptionally distinct cell populations. The
experiment was performed in 7 batches.

# Data availability

## SRA

The [Sequence Read Archive (SRA)](https://www.ncbi.nlm.nih.gov/sra) is the 
largest publicly available repository of high-throughput sequencing data.
The data are stored by the National Center for Biotechnology Information (NCBI) 
services and multiple cloud storage providers. From this website "raw" 
sequencing data can be retrieved. In practice, these are usually `.sra` files,
which can be downloaded and converted into FASTQ files using functions
from the `sratoolkit` software. 

For our dataset, the FASTQ data can be retrieved from this 
[link](https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA267857&o=acc_s%3Aa).
The data are stored as one file per sequencing batch, with each file 
approximately 20Gb. As such, it will be unfeasible to download and process these
FASTQ files in this practical session.

Instead, for demonstrative purposes, we have taken a subsample of the FASTQ
file for the first sequencing batch for you to work with. On this subsample, 
we may perform all the tasks that we would have performed on the full dataset.
The steps that are required for downloading and quantifying drop-seq data
can be found in [a Shell script on our companion GitHub repository](https://github.com/statOmics/singleCellCourse/blob/master/lab1_preprocessing/preprocessDropseq.sh).

## GEO

The dataset from Macosko *et al.* was also uploaded by the authors on the 
Gene Expression Omnibus (GEO) platform under accession number 
[GSE63472](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63472), 
from which raw and readily-processed data files may be retrieved, including:

1. *GSE63472_RAW.tar*, a 90.6Gb object that contains the "raw data" for the
experiment. In the scRNA-seq context, FASTQ files are often considered the raw
data format.

2. *GSE63472_P14Retina_merged_digital_expression.txt.gz*, a 50.7Mb matrix that
stores the gene expression values for each cell. These values are integer 
counts, that did not undergo any type of preprocessing or normalization.

3. *GSE63472_mm10_reference_metadata.tar.gz*, a 862.9Mb compressed folder
containing information on the reference genome to which the scRNA-seq reads
were aligned (see theory slides).

4. *GSE63472_P14Retina_logDGE.txt.gz*, a 316.8Mb compressed text file, not clear
what it contains (results from a differential gene expression analysis, but
with log-transformation, so log-fold changes maybe?).

As such, by downloading *GSE63472_P14Retina_merged_digital_expression.txt.gz*,
we avoid re-quantifying the data, i.e., the translation from reads from the
FASTQ files into a gene-level expression values for each cell. 

One issue that often arises from data downloaded from GEO, is that there is no
strict requirements for which data should be included in the upload by the 
authors. As such, from my personal experience, it can often be the case that
important information like metadata are missing, or the content of the submitted
files is unclear. Even if all the required data are available, as is the case 
here, we would still need to piece all the information together from different 
files and file formats before we can use them.

## ExperimentHub

The Bioconductor *ExperimentHub* web resource, which can be accessed using the 
[ExperimentHub](https://bioconductor.org/packages/release/bioc/html/ExperimentHub.html) 
R package, provides a central location where curated data from experiments, 
publications or training courses can be accessed. While it contains far less
datasets than the SRA or GEO (4965 records to date), these datasets all follow 
the tidy data format of Bioconductor. Note that the ExperimentHub contains 
several types of data, like bulk and single-cell transcriptomics data, 
microarrays and more.

The Macosko dataset is available from ExperimentHub and can be accessed as
follows:

```{r, eval=FALSE, message=FALSE, warning=FALSE}
library(ExperimentHub)
edb <- ExperimentHub()

edb[grep("Macosko", edb$title)] # find accession number (can be inefficient)

edb_counts <- edb[["EH2690"]]
edb_counts[1:5,1:5]

edb_coldata <- edb[["EH2691"]]
edb_coldata[1:5,]

rm(edb, edb_counts, edb_coldata)
```

## scRNASeq

In addition to ExperimentHub, Bioconductor provides the package
[scRNAseq](https://bioconductor.org/packages/release/data/experiment/html/scRNAseq.html).
This package provides an even more user-friendly client to access (only) 
scRNA-seq datasets from the ExperimentHub web resource. Data retrieved using the
scRNAseq package are stored as user-friendly `SingleCellExperiment` objects, 
with the expression data, gene-level information, cell-level information and 
experiment metadata all in place in one data object. The scRNA-seq package
currently holds 61 datasets, including the data from *Macosko et al.*:

```{r, eval=FALSE, message=FALSE, warning=FALSE}
library(scRNAseq)
scRNAseq::MacoskoRetinaData()
```

# Import data

The `scRNAseq` package provides convenient access to several datasets. See the [package Bioconductor page](http://bioconductor.org/packages/release/data/experiment/html/scRNAseq.html) 
for more information.

```{r, eval=FALSE}
# Code below might ask you to create an ExperimentHub directory. 
# Type 'yes' and hit Enter, to allow this.
suppressPackageStartupMessages(library(scRNAseq))
sce <- MacoskoRetinaData()
```

# A `SingleCellExperiment` object

```{r, eval=FALSE}
sce
```

## Accessing data from a `SingleCellExperiment` object

Please see [Figure 4.1 in OSCA](http://bioconductor.org/books/release/OSCA/data-infrastructure.html) 
for an overview of a `SingleCellExperiment` object.

```{r, eval=FALSE}
# Data: assays
assays(sce)
assays(sce)$counts[1:5, 1:5]

# Feature metadata: rowData
rowData(sce) # empty for now

# Cell metadata: colData
colData(sce)

# Reduced dimensions: reducedDims
reducedDims(sce) # empty for now
```

## Creating a new `SingleCellExperiment` object

```{r, eval=FALSE}
sceNew <- SingleCellExperiment(assays = list(counts = assays(sce)$counts))
sceNew

rm(sceNew)
```

## Storing (meta)data in a `SingleCellExperiment` object

```{r, eval=FALSE}
fakeGeneNames <- paste0("gene", 1:nrow(sce))
rowData(sce)$fakeName <- fakeGeneNames
head(rowData(sce))
# Remove again by setting to NULL
rowData(sce)$fakeName <- NULL

assays(sce)$logCounts <- log1p(assays(sce)$counts)
assays(sce)
assays(sce)$logCounts[1:5, 1:5]
assays(sce)$logCounts <- NULL
```

# Obtaining and including rowData

The `rowData` slot of a `SingleCellExperiment` object allows for storing 
information on the features, i.e. the genes, in a dataset. In our object,
the `rowData` slot is empty.

```{r, eval=FALSE}
rowData(sce)
```

As such, the only information we have on the genes are their names, which can
be retrieved as the `rownames` of the expression matrix.

```{r, eval=FALSE}
head(rownames(sce))
```

These are the gene names (symbols). Note that it may be useful to include 
additional information in the `rowData` slot. For instance, we may want to 
store:

- Unambiguous gene identifiers (e.g. from ENSEMBL)
- On which chromosome the gene is located
- Gene length (genomic start position and end position)
- Others...

```{r, eval=FALSE, message=FALSE, warning=FALSE}
library("biomaRt")

ensembl75 <- useEnsembl(biomart = 'genes', 
                        dataset = 'mmusculus_gene_ensembl',
                        version = 75)

head(listAttributes(ensembl75)) # potential info to extract

geneInfo <- getBM(attributes = c("ensembl_gene_id", # ENSEMBL unambiguous identifier
                                 "mgi_symbol", # Gene symbol (to link with SCE rownames),
                                 "chromosome_name", # On which chromosome
                                 "start_position", # Start position
                                 "end_position"),# End position
                  mart = ensembl75)

head(geneInfo)
```

```{r, eval=FALSE}
geneInfo$mgi_symbol_upper <- toupper(geneInfo$mgi_symbol) 
# match between gene info and rownames

sum(rownames(sce) %in% geneInfo$mgi_symbol_upper)
sum(!rownames(sce) %in% geneInfo$mgi_symbol_upper) # lost in conversion :(

rowData(sce) <- geneInfo[match(rownames(sce),geneInfo$mgi_symbol_upper),]
```

```{r, eval=FALSE}
head(rowData(sce))
```

# Filtering non-informative genes

```{r, eval=FALSE}
library(edgeR)

# A very simple strategy: remove all genes that are expressed in less than 10
# out of 49300 cells -> note that this a very lenient filtering criterium
keep <- rowSums(assays(sce)$counts > 0) > 10
table(keep)

sce <- sce[keep,]
```

Note that dedicated functions for filtering out lowly expressed genes exist.
One such function is the `filterByExpr` function of the edgeR R package.
In brief, the strategy keeps genes that have at least "min.count" reads in a 
worthwhile number samples. 

More precisely, the filtering keeps genes that have count-per-million (CPM) 
above k in n samples. 

k is determined by the min.count argument to the function, and by the sample 
library sizes.

n is determined by the design matrix. n can be seen as the smallest group sample
size. A `group` of samples/cells can be defined as cells that are more similar
to one another, e.g., from the same sequencing batch, the same patient....
Here we could also use the cluster assignment (cell type) for each cell as the
grouping variable; *note however, that this usually is not available until a*
*later stage in the analysis pipeline* (i.e., after dimension reduction and 
clustering, topics we will cover next session.)

If all the group sizes are larger than the `large.n` argument of the 
filterByExpr function, which defaults to 10, then n will be taken as
`min.prop`* the the number of samples/cells in the smallest group.

Note that all the group sizes will often be larger than the `large.n` in case
of single-cell data.

```{r, eval=FALSE}
# Slow (more than 1min) -> do not run
# keep2 <- filterByExpr(sce,
#                       group = sce$cluster,
#                       min.count = 1,
#                       min.total.count = 15,
#                       min.prop = 0.2)
# table(keep2)
```


# Quality control

In quality control (QC), we check the quality of our dataset. In particular, 
we investigate undesirable oddities, such as low-quality cells, empty droplets
or doublets.

## Identifying and removing low-quality cells

There are several distinguishing features of low-quality cells that can be used
in order to identify them. [As described in the OSCA book](http://bioconductor.org/books/3.14/OSCA.basic/quality-control.html) 
book):

1. The library size is defined as the total sum of counts across all 
relevant features for each cell. Here, we will consider the relevant features 
to be the endogenous genes. Cells with small library sizes are of low quality 
as the RNA has been lost at some point during library preparation, either due 
to cell lysis or inefficient cDNA capture and amplification.

2. The number of expressed features in each cell is defined as the number of 
endogenous genes with non-zero counts for that cell. Any cell with very 
few expressed genes is likely to be of poor quality as the diverse transcript 
population has not been successfully captured.

3. We sometimes have spike-in (ERCC) transcripts available.
The proportion of reads mapped to spike-in transcripts is calculated relative 
to the total count across all features (including spike-ins) for each cell. 
As the same amount of spike-in RNA should have been added to each cell, 
any enrichment in spike-in counts is symptomatic of loss of endogenous RNA.

4. In the absence of spike-in transcripts, the proportion of reads mapped 
to genes in the mitochondrial genome can be used. High proportions are 
indicative of poor-quality cells (Islam et al. 2014; Ilicic et al. 2016), 
presumably because of loss of cytoplasmic RNA from perforated cells. 
The reasoning is that, in the presence of modest damage, the holes in the 
cell membrane permit efflux of individual transcript molecules but are too 
small to allow mitochondria to escape, leading to a relative enrichment of 
mitochondrial transcripts. For single-nuclei RNA-seq experiments, high 
proportions are also useful as they can mark cells where the cytoplasm has 
not been successfully stripped.

## Calculate QC variables

This function calculates useful QC metrics for identification and removal of 
potentially problematic cells. Default per-cell metrics are the sum of counts 
(i.e., the library size) and the number of detected features. The percentage of 
counts in the top features also provides a measure of library complexity.

If subsets is specified, these statistics are also computed for each subset of 
features. This is useful for investigating gene sets of interest, e.g., 
mitochondrial genes.

```{r, eval=FALSE, message=FALSE, warning=FALSE}
library(scater)

# check ERCC spike-in transcripts
sum(grepl("^ERCC-", rownames(sce))) # no spike-in transcripts available

# check mitochondrial genes
sum(rowData(sce)$chromosome_name=="MT",na.rm = TRUE) # 28 mitochondrial genes
sum(grepl("^MT-", rownames(sce))) # alternatively
is.mito <- grepl("^MT-", rownames(sce))

## calculate QC metrics
df <- perCellQCMetrics(sce, subsets=list(Mito=is.mito))
head(df)

# add QC variables to sce object
colData(sce) <- cbind(colData(sce), df)

# the QC variables have now been added to the colData of our SCE object.
head(colData(sce))
```

## Exploratory data analysis

In the figure below, we see that several cells have a very low number of 
expressed genes, and where most of the molecules are derived from 
mitochondrial genes. This indicates likely damaged cells, presumably because 
of loss of cytoplasmic RNA from perforated cells, so we should remove these for 
the downstream analysis.

```{r, eval=FALSE}
# Number of genes vs library size
plotColData(sce, x = "sum", y="detected", colour_by="cluster") 

# Mitochondrial genes
plotColData(sce, x = "detected", y="subsets_Mito_percent")
```

## QC using adaptive thresholds

Below, we remove cells that are outlying with respect to

 1. A low sequencing depth (number of UMIs);
 2. A low number of genes detected;
 3. A high percentage of reads from mitochondrial genes.
 
 Here we will remove cells for QC based on **adaptive thresholds** related to
 the three points from above. Adaptive trhesholds are used as opposed to
 **fixed thresholds**. 
 
With fixed thresholds, we use fixed cut-off values for each cell to pass QC, 
e.g., we might consider cells to be low quality if they have library sizes 
below 100,000 reads; express fewer than 5,000 genes; have spike-in proportions
above 10%; or have mitochondrial proportions above 10%.

With adaptive thresholds, we assume that most of the dataset consists of 
high-quality cells. We then identify cells that are outliers for the various 
QC metrics, based on the median absolute deviation (MAD) from the median value
of each metric across all cells. By default, we consider a value to be anoutlier 
if it is more than 3 MADs from the median in the “problematic” direction. This 
is loosely motivated by the fact that such a filter will retain 99% of 
non-outlier values that follow a normal distribution. We demonstrate adopting
adaptive thresholds on the Macosko dataset:

```{r, eval=FALSE}
lowLib <- isOutlier(df$sum, type="lower", log=TRUE)
lowFeatures <- isOutlier(df$detected, type="lower", log=TRUE)
highMito <- isOutlier(df$subsets_Mito_percent, type="higher")

table(lowLib)
table(lowFeatures)
table(highMito)

discardCells <- (lowLib | lowFeatures | highMito)
table(discardCells)
colData(sce)$discardCells <- discardCells

# visualize cells to be removed
plotColData(sce, x = "sum", y="detected", colour_by="discardCells")
plotColData(sce, x = "detected", y="subsets_Mito_percent", colour_by = "discardCells")
```

We removed a total of $3423$ cells, most of which because of an outlyingly high 
percentage of reads from mitochondrial genes.

## Identifying and removing empty droplets

Note that the removal of cells with low sequencing depth using the adaptive 
threshold procedure above is a way of removing empty droplets. 
Other approaches are possible, e.g., removing cells by statistical testing 
using `emtpyDrops`. This does require us to specify a lower bound on the total 
number of UMIs, below which all cells are considered to correspond to empty 
droplets. This lower bound may not be trivial to derive, but the `barcodeRanks`
function can be useful to identify an elbow/knee point.

In brief, the steps taken by the `emtpyDrops` function can be summarized as 
follows:

1.	Define threshold T of total UMI counts (e.g. with the help of the 
`barcodeRanks` function), below which cells may be considered to be from empty 
droplets. Call this set of cells E.

2.	Define $A_g$ as the total gene expression across all cells in E.

3.	Define $pi_g$ as the relative contribution of gene g to the ambient profile.

4.	Calculate p-value for each cell to have a transcriptional profile similar to
the ambient solution. Intuitively, a p-value below the requested alpha level 
would correspond to a cell for which the observed count profile strongly 
deviates from the count profile observed in the cells with a library size 
below threshold T, i.e., a non-empty droplet.

```{r, eval=FALSE, message=FALSE, warning=FALSE}
library(DropletUtils)
bcrank <- barcodeRanks(counts(sce))

# Only showing unique points for plotting speed. Duplicated ranks are a 
# consequence of ties in the ranks, i.e., when cells have an equal library size.
sum(duplicated(bcrank$rank))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log="xy",
    xlab="Rank", ylab="Total UMI count", cex.lab=1.2)
abline(h=metadata(bcrank)$inflection, col="darkgreen", lty=2)
abline(h=metadata(bcrank)$knee, col="dodgerblue", lty=2)
abline(h=350, col="orange", lty=2) # picked visually myself
legend("topright", 
       legend=c("Inflection", "Knee", "Empirical knee point"), 
       col=c("darkgreen", "dodgerblue", "orange"), 
       lty=2, 
       cex=1.2)

set.seed(100)
limit <- 350   
all.out <- emptyDrops(counts(sce), lower=limit, test.ambient=TRUE)
# p-values for cells with total UMI count under the lower bound.
hist(all.out$PValue[all.out$Total <= limit & all.out$Total > 0],
    xlab="P-value", main="", col="grey80", breaks=20)

# but note that it would remove a very high number of cells
length(which(all.out$FDR <= 0.01)) # retained

# so we stick to the more lenient adaptive filtering strategy
# remove cells identified using adaptive thresholds
sce <- sce[, !colData(sce)$discardCells]
```

## Identifying and removing doublets

We will use 
[scDblFinder](https://bioconductor.org/packages/3.14/bioc/html/scDblFinder.html) 
to detect doublet cells.

As discussed in the theory session of last week, the steps taken by 
`scDblFinder` can be summarized as follows:

1. Perform principal components analysis (PCA) on log-normalized expression 
counts. This allows for projecting each cell in the dataset into a 2D space
(for more details on PCA, see next weeks session).

2. Randomly select two cells, sum their counts and normalize, and project 
into PCA space from step 1. In other words, artificially generate doublets and
see where they are located in the 2D space.

3. Repeat step 2 many times (generate many artificial doublets).

4. Generate neighbor network in the 2D space. The network is then used to 
estimate a number of characteristics for each cell, in particular the proportion 
f artificial doublets among the nearest neighbors.

5. Use this information, along with other predictors, to train a classifier
(gradient boosted tree) that allows for distinguishing doublets from singlets.

Note: only classifies for identifying doublets for which the two cells
are from different cell type clusters.


```{r, eval=FALSE, message=FALSE, warning=FALSE}
## perform doublet detection
library(scDblFinder)
```

```{r, eval=FALSE}
set.seed(211103)
sampleID <- unlist(lapply(strsplit(colData(sce)$cell.id, split="_"), "[[", 1))
table(sampleID)
sce <- scDblFinder(sce, 
                   returnType="table",
                   samples = factor(sampleID))
table(sce$scDblFinder.class)


## visualize these scores
## explore doublet score wrt original cluster labels
boxplot(log1p(sce$scDblFinder.score) ~ factor(colData(sce)$cluster, exclude=NULL))

tab <- table(sce$scDblFinder.class, sce$cluster, 
      exclude=NULL)
tab
t(t(tab) / colSums(tab))

barplot(t(t(tab) / colSums(tab))[2,],
        xlab = "Cluster", ylab = "Fraction of doublets")

range(sce$scDblFinder.score[sce$scDblFinder.class  == "doublet" & sampleID == "r1"])
range(sce$scDblFinder.score[sce$scDblFinder.class  == "singlet" & sampleID == "r1"])
```

```{r, eval=FALSE}
# remove doublets
sce <- sce[,!sce$scDblFinder.class == "doublet"]
```

# Normalization

Normalization aims to remove technical effects such as sequencing depth so that 
comparisons between cells are not confounded by them. The most commonly used 
normalization methods methods use scaling, where a scaling factor (also called 
size factor, normalization factor) is estimated for each cell. These scaling 
factors (e.g., the effective library size) can be included as an offset
to downstream modeling procedures. This effectively allows for performing
inference on the relative abundance of a gene in a cell, after accounting for 
library size and RNA composition differences between cells, which is much more
relevant than comparing raw counts

Note that we always prefer the use of offsets over transformation of the data.
In brief, such transformation would distort the mean-variance relationship
in the data. For more details, we refer to the document that was discussed
in theory session of last week
([link](https://statomics.github.io/SGA21/sequencing_scalingNormalization.html))

For normalization, the size factors $s_i$ computed here are simply scaled 
library sizes:
\[ N_i = \sum_g Y_{gi} \]
\[ s_i = N_i / \bar{N}_i \]

```{r, eval=FALSE}
sce <- logNormCounts(sce)

# note we also returned log counts: see the additional logcounts assay.
sce

# you can extract size factors using
sf <- librarySizeFactors(sce)
mean(sf) # equal to 1 due to scaling.
plot(x= log(colSums(assays(sce)$counts)), 
     y=sf)
```

From the OSCA book:
Alternatively, we may use more sophisticated approaches for variance stabilizing 
transformations in genomics data, e.g., `DESeq2` or `sctransform`. These aim 
to remove the mean-variance trend more effectively than the simpler 
transformations mentioned above, though it could be argued whether this is 
actually desirable. For low-coverage scRNA-seq data, there will always be a 
mean-variance trend under any transformation, for the simple reason that the 
variance must be zero when the mean count is zero. These methods also face the 
challenge of removing the mean-variance trend while preserving the interesting 
component of variation, i.e., the log-fold changes between subpopulations; this 
may or may not be done adequately, depending on the aggressiveness of 
the algorithm.

In practice, the log-transformation is a good default choice due to its 
simplicity and interpretability, and is what we will be using for all 
downstream analyses.

---

--- end lab session 1 ---

---

We have created a 
[DropBox folder](https://www.dropbox.com/sh/3yc9065ri76b2e7/AAAr0z9ZGp3APZ-reH6I0DGQa?dl=0) 
containing convenience files, e.g., a preprocessed `SingleCellExperiment` that 
is the end result of lab session 1. You may want to downlaod them and import them in the code chunk below.

```{r, eval=FALSE}
```

# Normalization (continued)

Normalization is necessary to correct for several sources of technical 
variation:

 - **Differences in sequencing depth** between samples. Some samples get 
 sequenced deeper in the sense that they consist of more (mapped) reads and 
 therefore can be considered to contain a higher amount of information, which 
 we should be taking into account. In addition, if a sample is sequenced deeper, 
 it is natural that the counts for each gene will be higher, jeopardizing a 
 direct comparison of the expression counts.
 - **Differences in RNA population composition** between samples. As an extreme 
 example, suppose that two samples have been sequenced to the exact same depth.
 One sample is contaminated and has a very high concentration of the 
 contaminant cDNA being sequenced, but otherwise the two samples are identical. 
 Since the contaminant will be taking up a significant proportion of the reads 
 being sequenced, the counts will not be directly comparable between the 
 samples. Hence, we may also want to correct for differences in the composition 
 of the RNA population of the samples
 (see [edgeR manual chapter 2.8](https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)).
 - **Other technical variation** such as sample-specific GC-content or 
 transcript length effects may also be accounted for.

```{r, eval=FALSE}
table(sce$cluster)
```

Below, we will visualize the normalization occurring between two cells of the 
same cell type (which could be considered technical repeats): 

```{r, eval=FALSE}
select <- sce$cluster == "1"
select[is.na(select)] <- FALSE
cs <- colSums(assays(sce)$counts[,select])
cs[order(cs, decreasing = TRUE)][c(1,10)]
```

Let’s take a look at how comparable two cells (replicates) of cluster 1 are. 
We will compare the cell with the highest library size with the cell that has 
the 10th highest library size using MD-plots (mean-difference plots, as 
introduced by 
[Dudoit et al. (2002)](https://www.jstor.org/stable/24307038?seq=1#metadata_info_tab_contents)), 
also sometimes referred to as MA-plots.

```{r, eval=FALSE}
targetCells <- names(cs[order(cs, decreasing = TRUE)][c(1,10)])

M <- rowMeans(assays(sce)$counts[,targetCells])
D <- assays(sce)$counts[,targetCells[2]] / assays(sce)$counts[,targetCells[1]]
plot(x = log(M), y = log2(D),
     pch = 16, cex=1/3,
     main = paste0("Cell ", targetCells[2], " vs cell ", targetCells[1]),
     xlab = "Log mean", ylab = "Log2 fold-change",
     bty = 'l')
abline(h = 0, col="orange", lwd=2)

```

We see clear bias in the comparison of the 1st and 10th most deeply sequenced 
cell from cell cluster 1. We see that the log fold-changes are biased downwards. 
This means that, on average, a gene is higher expressed in cell 1 versus cell 10.
Looking at the library sizes, we can indeed see that the library size for 
cell 1 is 20869 counts, while it is only 14719 counts for cell 10! This is a
clear library size effect that we should take into account.

```{r, eval=FALSE}
# normalize the count data using the previously computed "size factors"
assay(sce, "normed") <- normalizeCounts(sce, 
                                        log=FALSE,
                                        size.factors=sf, 
                                        pseudo.count=0)

M <- rowMeans(assays(sce)$normed[,targetCells])
D <- assays(sce)$normed[,targetCells[2]] / assays(sce)$normed[,targetCells[1]]
plot(x = log(M), y = log2(D),
     pch = 16, cex=1/3,
     main = paste0("Cell ", targetCells[2], " vs cell ", targetCells[1]),
     xlab = "Log mean", ylab = "Log2 fold-change",
     bty = 'l')
abline(h = 0, col="orange", lwd=2)
```

Upon normalizing the data using size factors, we have removed bias as a
consequence of differences in sequencing depth.

Note that we computed the normalized count matrix only for demonstrating the
effect of such normalization. In a typical workflow, the size factors are used
as offsets in the downstream models rather than to perform data transformation.
In brief, such transformation would distort the mean-variance relationship
in the data. For more details, we refer to the document that was touched upon
in the first theory session
([link](https://statomics.github.io/SGA21/sequencing_scalingNormalization.html))

# Feature selection

As dimensions increase, shortest and farthest distances between points become
nearly inseparable. In high-dimensional space, it is therefore extremely 
difficult to separate signal from noise.

In order to recover structure (e.g. setting up a dimension-reduced
space to help us find cell-type clusters in the data), we want to move to an
informative, lower-dimensional space. We will select genes which we hope are
informative for recovering the biological structure. But what defines an 
informative gene?

## Selecting genes with high variance

The simplest approach to quantifying per-gene variation is to compute the 
variance of the log-normalized expression values (i.e., "log-counts") for 
each gene across all cells. 

```{r,eval=FALSE}
# calculate variance of log-normalized counts for each gene
geneVars <- genefilter::rowVars(...)
# select top 1000 highly variable genes
highVarGenes <- names(geneVars)[order(geneVars, decreasing=TRUE)[1:1e3]]
head(highVarGenes)
```

## Selecting genes with high variation with respect to mean

While calculation of the per-gene variance is simple, feature selection requires
modelling of the mean-variance relationship. Indeed, not accounting for the 
mean-variance structure while selecting highly variable genes will oftentimes 
boil down to selecting the most highly expressed genes. The log-transformation 
is a helpful variance stabilizing transformation, however, it is not perfect, 
meaning that the variance of a gene is not completely independent of its mean.
Therefore, feature selection may still be driven by average expression rather 
than underlying biological heterogeneity.

A (rightly so) popular approach is to select genes that have a high variance
with respect to their mean. Often, first an empirical mean-variance trend is
fitted, upon which genes with the highest positive residuals are selected. 
Being intuitive, reasonable and fairly straight-forward, this method is widely
used.

To account for the mean-variance effect, we use the `modelGeneVar` function of
the `scran` package to fit a trend to the variance with respect to abundance
across all genes (on log-normalized expression values of the sce object).

```{r,eval=FALSE}
library(scran)
# `modelGeneVar` function description: Model the variance of the log-expression 
# profiles for each gene, decomposing it into technical and biological 
# components based on a fitted mean-variance trend.
dec <- modelGeneVar(...) # input a sce object with precomputed logcounts assay
head(dec)
```

The fitted value for each gene is used as a proxy for the technical component of
variation for each gene, under the assumption that most genes exhibit a low
baseline level of variation that is not biologically interesting. The biological
component of variation for each gene is defined as the the residual from the
trend. Ranking genes by the biological component enables identification of
interesting genes for downstream analyses in a manner that accounts for the
mean-variance relationship.

```{r,eval=FALSE}
fitRetina <- metadata(...)
plot(x = ..., # the fitted means on the x-axis 
     y = ..., # the fitted variances on the y-axis
     xlab="Mean of log-expression",
    ylab="Variance of log-expression")
curve(fitRetina$trend(x), col="dodgerblue", add=TRUE, lwd=2) # adds a trend line
```

We are interested in those genes for which the variance in expression is higher
than what we would expect for that gene based on its mean expression.

```{r,eval=FALSE}
# get 10% most variable genes
# Based on the output statistics of modelGeneVar, select the 10% most variable genes
hvg <- getTopHVGs(stats = ..., 
                  prop = ...) # percentage of top variable genes. Alternatively,
# n = ... could be used to select the top number of variable genes
head(hvg)

# plot these 
plot(x = ..., # the fitted means on the x-axis 
     y = ..., # the fitted variances on the y-axis
     col = c("orange", "darkseagreen3")[(names(fitRetina$mean) %in% hvg)+1],
     xlab="Mean of log-expression",
    ylab="Variance of log-expression")
curve(fitRetina$trend(x), col="dodgerblue", add=TRUE, lwd=2)
legend("topleft", 
       legend = c("Selected", "Not selected"), 
       col = c("darkseagreen3", "orange"),
       pch = 16,
       bty='n')
```

As a comparison, we could color the genes on this figure according to the 
selection that was made purely by looking at the raw variance of each gene.

```{r,eval=FALSE}
...
```

As expected, we here simply select genes with a high variance, without 
recognizing that a high variance is typically driven by a high mean.

## High deviance genes

Here, we will select genes with a high residual deviance. The idea is that we 
assume a null model of constant expression fraction (i.e., RNA concentration) 
across all cells. We subsequently calculate a goodness-of-fit statistic for 
each gene, assessing whether the model is a good approximation to the gene 
expression of the corresponding gene.

If the model fits poorly, i.e., the gene has a high deviance, the expression 
fraction varies significantly across the cells in our datasets.Genes with a 
high deviance will thus most poorly fit a null model where the relative
abundance is equal for all cells, which therefore are informative.

```{r, eval=FALSE}
#BiocManager::install("scry")
library(scry)
# `devianceFeatureSelection` function description: Computes a deviance statistic 
# for each row feature (such as a gene) for count data based on a multinomial 
# null model [...]. 
sce <- devianceFeatureSelection(object = ..., 
                                assay = ...) # computed on raw counts

plot(sort(rowData(sce)$binomial_deviance, decreasing = TRUE), 
     type="l", 
     xlab="ranked genes", 
     ylab="binomial deviance", 
     main="Feature Selection with Deviance")
abline(v=2000, lty=2, col="red")
```

Our plot looks similar to one displayed in the 
[vignette of the scry package](https://www.bioconductor.org/packages/release/bioc/vignettes/scry/inst/doc/scry.html). 
Based on that plot, the authors suggest retaining 2.000 genes (the top 2000 
based on the deviance residuals) for downstream dimensionality reduction and 
clustering.

## Seurat VST

Another very common feature selection strategy is the variance-stabilizing
transformation from the `Seurat` R package, which amounts to calculating Pearson
residuals from a regularized negative binomial regression model, with sequencing 
depth as a covariate.

**Intermezzo: interoperability between SingleCellExperiment and Seurat objects**

In this lecture series, we always make use of the SingleCellExperiment object
and the packages available from Bioconductor. Another very popular toolbox
for performing scRNA-seq data analysis is `Seurat`. However, functions from
`Seurat` cannot be used directly to manipulate `SingleCellExperiment` objects,
and vice versa. Fortunately, efforts have been made to increase the
interoperability between the two toolboxes.

```{r, eval=FALSE}
library(Seurat)
seurat_obj <- as.Seurat(...) # sce object
seurat_obj # notice the "0 variable features"
```

On this object, we may use functions from the `Seurat` toolbox.
For instance, we may search for highly variable features using `Seurat`'s VST
implementation:

```{r, eval=FALSE}
# Prior to detecting variable features using VST, Seurat normalizes the data.
# The `NormalizeData` function takes as input the Seurat object, and uses 
# log normalization and a scale factor of 10000 as default (see help file).
seurat_obj <- Seurat::NormalizeData(object = ..., 
                                    normalization.method = ..., 
                                    scale.factor = ...)

# Next, perform feature selection based on VST
seurat_obj <-  FindVariableFeatures(object = ...,
                                    selection.method = ...)
```

```{r,eval=FALSE}
seurat_obj  # notice the "2000 variable features" (default)
head(VariableFeatures(seurat_obj)) # here they are
```

# Dimensionality reduction

Note that, below, we color the cells using the known, true cell type label as 
defined in the metadata, to empirically evaluate the dimensionality reduction. 
In reality, we don't know this yet at this stage.

## The most basic DR

Just by looking at the top two genes based on our feature selection criterion, 
we can already see some separation according to the cell type!

```{r,eval=FALSE}
colData(sce)$cluster <- as.factor(colData(sce)$cluster)
cl <- colData(sce)$cluster
par(bty='l')
plot(x = ..., # extract the counts for the most variable gene
     y = ..., # extract the counts for the second most variable gene
     col = as.numeric(cl),
     pch = 16, cex = 1/3,
     xlab = "Most informative gene",
     ylab = "Second most informative gene",
     main = "Cells colored acc to cell type")
```

We are able to recover quite some structure. However, many cell populations 
remain obscure, and the plot is overcrowded.

## Linear dimensionality reduction: PCA

A DR method is linear when the reduced dimensions are a linear function of the
original variables. For example, in PCA, each principal component is a linear
combination of genes, therefore the DR is a linear function of the original
variables.

Typically, PCA is performed on log-transformed normalized counts. 
The log-transformation helps  somewhat, but not completely, to account for the
mean-variance relationship. PCA works well for bulk RNA-seq data. However, the
structure of scRNA-seq data is often too complex to be visualized by a small 
number of PCs.

There are several R functions that allow you to perform PCA. Here, we make use
of the `runPCA` function of the `scater` package, which has been specifically
developed for performing PCA on `SingleCellExperiment` objects.
We calculate the top 30 principal components.

### PCA with feature selection

```{r,eval=FALSE}
# Perform a PCA for our data. Compute the first 30 principal components, only
# using the selected features with high variance with respect to the mean.
set.seed(1234)
sce <- runPCA(...) # you need three arguments
```

PCA has been performed. The PCA information has been automatically stored in the
*reducedDim* slot of the SingleCellExperiment object.

```{r,eval=FALSE}
reducedDimNames(sce)
```

```{r,eval=FALSE}
head(reducedDim(sce,
                type="PCA"))
```

The `plotPCA` function of the `scater` package now allows us to visualize
the cells in PCA space, based on the PCA information stored in our object:

```{r,eval=FALSE}
plotPCA(sce, 
        colour_by = "cluster")
```

While the large number of clusters in this dataset makes it hard to 
distinguish between all the different colors, we can already see that PCA
retrieves some, but not all, of the structure in the data that was discovered
by the original authors.

How many of the top PCs should we retain for downstream analyses? The choice of
the number of PCs is a decision that is analogous to the choice of the number of
HVGs to use. Using more PCs will retain more biological signal at the cost of
including more noise that might mask said signal. On the other hand, using fewer
PCs will introduce competition between different factors of variation, where
weaker (but still interesting) factors may be pushed down into lower PCs and
inadvertently discarded from downstream analyses.

Most analysts will simply aim to use a “reasonable” but arbitrary value,
typically ranging from 10 to 50. This is often satisfactory as the later PCs
explain so little variance that their inclusion or omission has no major effect.

```{r, eval = FALSE}
percent.var <- attr(reducedDim(sce), "percentVar")
plot(percent.var, log="y", xlab="PC", ylab="Variance explained (%)")
plot(cumsum(percent.var), xlab="PC", ylab="Cumulative variance explained (%)")
```

Here, retaining ±15PCs seems reasonable. If you really prefer a more data-driven
way for determining this, there are procedures available that aim to 
computationally identify the elbow/knee point in the variance explained per 
PC plot.

```{r, eval = FALSE}
library(PCAtools)
chosen.elbow <- findElbowPoint(percent.var)
chosen.elbow
```

### PCA without feature selection

Note: more features -> computationally more intensive!

```{r, eval = FALSE}
set.seed(1234)
sceNoFS <- runPCA(x = ..., 
                  ncomponents = ..., 
                  subset_row = 1:nrow(sce))
plotPCA(...)
rm(sceNoFS) # remove the object, we don't need it anymore
```

While we use more information to make this PCA plot (17.887 genes) as compared
to the feature selected PCA plot (642 genes), we seem to retrieve less structure
in the data. This is the power of feature selection, an increase in the
signal-to-noise ratio!

### Effect of feature selection on PCA

First, we compare the different feature selection criteria, using the top 1000
highly variable genes for each method.

```{r, eval = FALSE}
# Get top 1000 highly variable features using the highly variable genes
# (w.r.t. mean), high deviance genes and the VST strategy of Seurat

hvg1000 <- ... # extract top 1000 highly variable genes
hdg1000 <- ... # extract top 1000 genes with high deviance
vst1000 <- ... # extract top 1000 genes based on VST analysis

# HVG strategy
plotPCA(
    runPCA(sce,
           ncomponents = 2, 
           subset_row = hvg1000), 
    colour_by = "cluster") +
  ggtitle("Highly variable genes")

# HDG strategy
plotPCA(
    runPCA(sce,
           ncomponents = 2, 
           subset_row = hdg1000), 
    colour_by = "cluster") +
  ggtitle("High-deviance genes")

# VST strategy
plotPCA(
    runPCA(sce,
           ncomponents = 2, 
           subset_row = vst1000), 
    colour_by = "cluster",
    ) +
  ggtitle("Seurat VST")
```

Next, we assess the sensitivity on the number of top features for the highly
variable genes method;

```{r, eval = FALSE}
hvg_all <- getTopHVGs(dec)

hvg2000 <- ... # extract top 2000 highly variable genes
hvg1000 <- ... # extract top 1000 highly variable genes
hvg100 <- ... # extract top 100 highly variable genes
hvg10 <- ... # extract top 10 highly variable genes
hvg5 <- ... # extract top 5 highly variable genes

plotPCA(
    runPCA(sce,
           ncomponents = 2), 
    colour_by = "cluster") +
  ggtitle("All genes")

plotPCA(
    runPCA(sce,
           ncomponents = 2, 
           subset_row = hvg2000), 
    colour_by = "cluster") +
  ggtitle("Top 2000 genes")

plotPCA(
    runPCA(sce,
           ncomponents = 2, 
           subset_row = hvg1000), 
    colour_by = "cluster") +
  ggtitle("Top 1000 genes")

plotPCA(
    runPCA(sce,
           ncomponents = 2, 
           subset_row = hvg100), 
    colour_by = "cluster") +
  ggtitle("Top 100 genes")

plotPCA(
    runPCA(sce,
           ncomponents = 2, 
           subset_row = hvg10), 
    colour_by = "cluster") +
  ggtitle("Top 10 genes")

plotPCA(
    runPCA(sce,
           ncomponents = 2, 
           subset_row = hvg5), 
    colour_by = "cluster") +
  ggtitle("Top 5 genes")
```

### Remarks on PCA

Visualizations of reduced dimensions from linear dimensionality reduction
methods are often "overcrowded", and it is hard to see structure (e.g., the PCA
plot we just saw). Non-linear dimensionality reduction methods can overcome this
problem. As the name suggests, the reduced dimensions are a non-linear function
of the observed data. We will not go into detail as to how these work under the
hood, but provide a few guidelines for the most popular methods. Often, the top
(~10-50) PCs are provided as input.

## A generalization of PCA for exponential family distributions.

PCA is implicitly based on Euclidean distances, corresponding to maximizing a
Gaussian likelihood, which is inappropriate for count data such as scRNA-seq.
[Townes et al.](https://doi.org/10.1186/s13059-019-1861-6) (2019) develop
GLM-PCA, a generalization of PCA for exponential family likelihoods. They posit,
using negative control data, that the data generative mechanism of UMI count
data can be considered to be multinomial.

The GLM-PCA strategy is implemented in the `glmpca` function of the `glmpca`
package.

Note that this function is quite computationally intensive. For regular PCA 
on log-transformed normalized counts, the underlying computations can be 
strongly simplified. Here, we work with the raw counts, which we assume to 
be *Poisson* distributed, requiring an iterative optimization scheme.

```{r, eval=FALSE}
# runs 2min on my laptop
library(glmpca)
set.seed(211103)
poipca <- glmpca(Y = ..., # counts for the higly variable genes
                 L = ..., # two latent dimensions (components)
                 fam = ..., # assume Poisson for droplet data
                 minibatch = "stochastic") # used in the package vignette by 
# the authors, but beyond the scope of this session.
reducedDim(sce, "PoiPCA") <- poipca$factors
plotReducedDim(...) # make sure to specify the right value for the `dimred` argument
```

Alternatively, we could adopt the GLM-PCA strategy using only the genes with 
high deviance.

```{r, eval = FALSE}
# Based on the diagnostic plot from the feature selection, we would select the 
# top 2000 genes with highest deviance. However, to reduce the running time of
# the function, I here select the top 500 (which still is quite slow - 5min).
hdg500 <- names(sort(rowData(sce)$binomial_deviance, decreasing = TRUE))[1:500]

...
```

The authors of the `glmpca` package note that GLM-PCA can be slow for large 
datasets. Therefore, they have implemented a fast approximation of the 
algorithm, which first fits a null model of constant expression for each gene 
across all cells, and subsequently fits standard PCA to either the Pearson or 
deviance residuals from the null model.

However, at least for me the `nullResiduals` function was **extremely slow**
even slower than the `glmpca` code above. Therefore, **I suggest not running**
this code, but I leave it in for reference.

```
sce <- nullResiduals(sce, assay="counts", type="deviance")
sce <- nullResiduals(sce, assay="counts", type="pearson")

pca<-function(Y, L=2, center=TRUE, scale=TRUE){
    #assumes features=rows, observations=cols
    res<-prcomp(as.matrix(t(Y)), center=center, scale.=scale, rank.=L)
    factors<-as.data.frame(res$x)
    colnames(factors)<-paste0("dim", 1:L)
    factors
}
pca_d <- pca(assay(sce[hdg,], "binomial_deviance_residuals"))
pca_d$resid_type <- "deviance_residuals"
pca_p <- pca(assay(sce[hdg,], "binomial_pearson_residuals"))
pca_p$resid_type <- "pearson_residuals"
cm <- as.data.frame(colData(sce[hdg,]))
pd <- rbind(cbind(cm, pca_d), cbind(cm, pca_p))
ggplot(pd, aes(x=dim1, y=dim2, colour=phenoid)) + geom_point() +
  facet_wrap(~resid_type, scales="free", nrow=2) +
  ggtitle("PCA applied to null residuals of high deviance genes")
```

## Non-linear dimensionality reduction: t-SNE

t-SNE focuses on preserving local rather than global distances. Therefore,
distances on a t-SNE reduced dimension plot can only be interpreted locally,
i.e., cells that are close together in reduced dimension will have a similar
transcriptome, but cells that are far away may not necessarily have a very distinct transcriptome.

Running t-SNE on a `SingleCellExperiment` object can be achieved with the
`runTSNE` function of the `scater` package. By default, this function will first
perform PCA, and use the top 50 PCs as an input to the actual t-SNE
algorithm. Since we already performed PCA, we may set `dimred = "PCA"` as
argument to the function. As such we will be performing a T-SNE on the 30 PCs
we computed before. If we would like to run t-SNE only using the top 10
PCs, we could set `n_dimred = 10`.

In addition, we may wish to set `external_neighbors=TRUE`, which increases the
speed of the algorithm for large datasets by applying a heuristic.

**Note:** for me this was the slowest function of the analysis (so far). If you
feel like your PC/laptop had a lot of trouble with the previous step(s), you
may consider not running this code, or running it on a subset of the data
(e.g., randomly subsampling cells) for demonstrational purposes. 
Alternatively, you may consider reducing the input space for the t-SNE 
algorithm, e.g. by setting `n_dimred = 5`.

```{r, eval = FALSE}
# (For me it takes 3min30 with n_dimred = 5)
sce <- runTSNE(x = ..., 
               dimred = ...,
               n_dimred = ...,
               external_neighbors = ...)
plotTSNE(sce,
         colour_by = "cluster")
```

## Non-linear dimensionality reduction: UMAP

It is often suggested that UMAP is better than t-SNE in preserving global 
differences. Therefore, UMAP is also often used in analyses such as trajectory 
inference, where this is important.

Running UMAP on a `SingleCellExperiment` object can be achieved with the
`runUMAP` function of the `scater` package.

```{r, eval = FALSE}
# Using top 10% highly variable genes and top 30 PCs
sce <- runUMAP(...)
plotUMAP(...)

# Using top 10% highly variable genes and top 10 PCs
plotUMAP(runUMAP(...),
  colour_by = "cluster")
```

</div>
<div class="footer">
    <hr>
    This work is licensed under the <a href= "https://creativecommons.org/licenses/by-nc-sa/4.0">
    CC BY-NC-SA 4.0</a> licence.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("lab2_MacoskoTemplate.Rmd");
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
