---
title: 'Lab3: Clustering, marker gene detection and cell type annotation'
author: "Koen Van den Berge and Jeroen Gilis"
date: "7/12/2021"
output:
    html_document:
      code_download: true    
      theme: cosmo
      toc: true
      toc_float: true
      highlight: tango
      number_sections: true
---

In this third lab session, we will perform clustering, marker gene detection and 
cell type annotation on the dataset from Macosko *et al.*.

# Macosko dataset

In the first lab session (24 November, 2021), we have quantified and 
pre-processed the droplet single-cell RNA-seq dataset (**drop-seq protocol**) 
from the publication by Macosko *et al.*, Cell 161, 1202â€“1214
[(link)](https://doi.org/10.1016/j.cell.2015.05.002). In this experiment, 
Macosko *et al.* sequenced 49,300 cells from the mouse retina, identifying 
39 transcriptionally distinct cell populations. The experiment was performed in
7 batches.

In the previous two session, we already performed several steps:

1. Import the Macosko dataset as `SingleCellExperiment` object from the 
`scRNAseq` Bioconductor package.

2. Include ENSEMBL gene identifiers in the `rowData`

3. Remove very lowly expressed genes

4. Remove low quality cells 
  4.1. Cells with outlying library size 
  4.2. Cells with outlying transcriptome complexity 
  4.3. Cells with outlying percentage of mitochondrial reads 
  4.4. Empty droplets 
  4.5. Doublets

5. Normalization 
  5.1. Compute log-normalized counts 
  5.2. Compute scaling factor to correct for differences in library size

6. Feature selection 
  6.1. Genes with high variance 
  6.2. Genes with high variance with respect to their mean expression 
  6.3. Genes with high deviance 
  6.4. Genes with high variance after variance-stabilizing transformation (VST)

7. Dimensionality reduction 
  7.1. Based on two most variable genes from step 6.2. 
  7.2. PCA 
  7.3. GLM-PCA 
  7.4. T-SNE 
  7.5. UMAP
 
 During this session, we will add the following steps to this workflow :
 
 - Clustering (graph-based, k-means and hierarchical clustering)
 - Marker gene detection
 - Cell type annotation (supervised and semi-supervised)
 
To guide you with these next steps, we provide with an Rmarkdown template
that you can fill out:

[lab3_macoskoTemplate](./lab3_macoskoTemplate.html)

We also provide the solution to this exercise here:

[lab3_macoskoWorkflow](./lab3_macoskoWorkflow.html)


---

