<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Koen van den Berge and Jeroen Gilis" />


<title>Preprocessing and quality control of scRNA-seq data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>


<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Single-cell course</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="introSequencingBulk.html">A brief introduction to sequencing technology</a>
    </li>
    <li>
      <a href="singleCellProtocols.html">Two popular single-cell protocols</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Labs
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="labSession_1.html">Lab 1</a>
    </li>
    <li>
      <a href="labSession_2.html">Lab 2</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/statOmics/Rmd-website">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://statomics.github.io/">statOmics</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
<li role="separator" class="divider"></li>
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Preprocessing and quality control of scRNA-seq data</h1>
<h4 class="author">Koen van den Berge and Jeroen Gilis</h4>
<h4 class="date">24/11/2021</h4>

</div>


<p>In this first lab session, we will perform quality control, gene-level quantification and data pre-processing on two different scRNA-seq datasets.</p>
<div id="macosko-dataset" class="section level1">
<h1><span class="header-section-number">1</span> Macosko dataset</h1>
<p>In this workshop session, we will preprocess the single-cell RNA-seq dataset from the publication by Macosko <em>et al.</em>, Cell 161, 1202–1214 from 2015 <a href="https://doi.org/10.1016/j.cell.2015.05.002">(link)</a>. This is the manuscript in which the droplet scRNA-seq technology <strong>Drop-seq</strong> was introduced. Six years after the original publication, drop-seq is still one of the most commonly adopted scRNA-seq protocols, as evidenced by the large number of citations for Macosko <em>et al.</em> (4.303 citations at November 3, 2021).</p>
<p>In this particular experiment, Macosko <em>et al.</em> sequenced 49,300 cells from the mouse retina, identifying 39 transcriptionally distinct cell populations. The experiment was performed in 7 batches.</p>
<p>The raw data (FASTQ files) of the experiment can be retrieved from the sequencing read archive (SRA) under the accession <a href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1853178">SRR1853178</a>. However, given the size of the data, we have provided a more amenable subset of the data on our lab GitHub page.</p>
<div id="data-quantification-with-qc" class="section level2">
<h2><span class="header-section-number">1.1</span> Data quantification with QC</h2>
<p>We will make use of <a href="https://salmon.readthedocs.io/en/latest/alevin.html">alevin</a> software, which is integrated with the <a href="https://salmon.readthedocs.io/en/latest/salmon.html">salmon</a> quantification software to quantify a subset of reads form the Macosko experiment, while simulataneously providing some quality control for the sequenced reads.</p>
<p>These steps are implemented in a small .shell script, which can be obtained using the following link <a href="./lab1_preprocessing/alevin_macosko/preprocessDropseq_lab.sh">Macosko_sh</a>.</p>
</div>
<div id="data-exploration-and-preprocessing-in-r" class="section level2">
<h2><span class="header-section-number">1.2</span> Data exploration and preprocessing in R</h2>
<p>Here, we provide an example of a script for performing data exploration and preprocessing in R for the Macosko dataset</p>
<p><a href="./lab1_MacoskoWorkflow.html">lab1_MacoskoWorkflow</a></p>
</div>
</div>
<div id="cuomo-dataset" class="section level1">
<h1><span class="header-section-number">2</span> Cuomo dataset</h1>
<p>We here make use of the publication of Anna Cuomo et al. (last author Oliver Stegle), which we will refer to as the <code>Cuomo dataset</code>. The paper that describes this dataset can be found using this <a href="https://www.nature.com/articles/s41467-020-14457-z">link</a>.</p>
<p>In the experiment, the authors harvested induced pluripotent stem cells (iPSCs) from 125 healthy human donors. These cells were used to study the endoderm differentiation process. In this process, iPSCs differentiate to endoderm cells, a process which takes approximately three days. As such, the authors cultered the iPSCs cell lines and allowed for differentiation for three days. During the experiment, cells were harvested at four different time points: day0 (directly at to incubation), day1, day2 and day3. Knowing the process of endoderm differentiation, these time points should correspond with different cell types: day0 are (undifferentiated) iPSCs, day1 are mesendoderm cells, day2 are “intermediate” cells and day3 are fully differentiated endoderm cells.</p>
<p>This dataset was generated using the <strong>SMART-Seq2</strong> scRNA-seq protocol.</p>
<p>The final goal of the experiment was to characterize population variation in the process of endoderm differentiation.</p>
<div id="data-quantification" class="section level2">
<h2><span class="header-section-number">2.1</span> Data quantification</h2>
<p>We will make use of the <a href="https://salmon.readthedocs.io/en/latest/salmon.html">salmon</a> quantification software to quantify a subset of the cells form the Cuomo experiment. Note that here we do not make use of Alevin, as Alevin is only used for drople-based data such as drop-seq and 10X data.</p>
<p>The salmon quantification is implemented in a small .shell script, which can be obtained using the following link <a href="./lab1_preprocessing/salmon_cuomo/quantify_cuomo_lab.sh">Cuomo_sh</a>.</p>
</div>
<div id="data-exploration-and-preprocessing-in-r-1" class="section level2">
<h2><span class="header-section-number">2.2</span> Data exploration and preprocessing in R</h2>
<p>Here, we provide a template for a script for performing data exploration and preprocessing in R for the Cuomo dataset. It is up to you to fill out this script, based on the concepts discussed in the theory session and the fucntions explained for the Macosko analysis!</p>
<p><a href="./lab1_CuomoTemplate.html">lab1_CuomoTemplate</a></p>
<p>Here, we provide a “solution” file for the analysis of the Cuomo dataset for the first lab session:</p>
<p><a href="./lab1_preprocessing/lab1_CuomoWorkflow.html">lab1_CuomoWorkflow_html</a> <a href="./lab1_preprocessing/lab1_CuomoWorkflow.Rmd">lab1_CuomoWorkflow_rmd</a></p>
<hr />
</div>
</div>

<div id="rmd-source-code">LS0tCnRpdGxlOiAiUHJlcHJvY2Vzc2luZyBhbmQgcXVhbGl0eSBjb250cm9sIG9mIHNjUk5BLXNlcSBkYXRhIgphdXRob3I6ICJLb2VuIHZhbiBkZW4gQmVyZ2UgYW5kIEplcm9lbiBHaWxpcyIKZGF0ZTogIjI0LzExLzIwMjEiCm91dHB1dDoKICAgIGh0bWxfZG9jdW1lbnQ6CiAgICAgIGNvZGVfZG93bmxvYWQ6IHRydWUgICAgCiAgICAgIHRoZW1lOiBjb3NtbwogICAgICB0b2M6IHRydWUKICAgICAgdG9jX2Zsb2F0OiB0cnVlCiAgICAgIGhpZ2hsaWdodDogdGFuZ28KICAgICAgbnVtYmVyX3NlY3Rpb25zOiB0cnVlCi0tLQoKSW4gdGhpcyBmaXJzdCBsYWIgc2Vzc2lvbiwgd2Ugd2lsbCBwZXJmb3JtIHF1YWxpdHkgY29udHJvbCwgZ2VuZS1sZXZlbCAKcXVhbnRpZmljYXRpb24gYW5kIGRhdGEgcHJlLXByb2Nlc3Npbmcgb24gdHdvIGRpZmZlcmVudCBzY1JOQS1zZXEgZGF0YXNldHMuCgojIE1hY29za28gZGF0YXNldAoKSW4gdGhpcyB3b3Jrc2hvcCBzZXNzaW9uLCB3ZSB3aWxsIHByZXByb2Nlc3MgdGhlIHNpbmdsZS1jZWxsIFJOQS1zZXEgZGF0YXNldApmcm9tIHRoZSBwdWJsaWNhdGlvbiBieSBNYWNvc2tvICpldCBhbC4qLCBDZWxsIDE2MSwgMTIwMuKAkzEyMTQgZnJvbSAyMDE1ClsobGluayldKGh0dHBzOi8vZG9pLm9yZy8xMC4xMDE2L2ouY2VsbC4yMDE1LjA1LjAwMikuIFRoaXMgaXMgdGhlIG1hbnVzY3JpcHQgaW4Kd2hpY2ggdGhlIGRyb3BsZXQgc2NSTkEtc2VxIHRlY2hub2xvZ3kgKipEcm9wLXNlcSoqIHdhcyBpbnRyb2R1Y2VkLgpTaXggeWVhcnMgYWZ0ZXIgdGhlIG9yaWdpbmFsIHB1YmxpY2F0aW9uLCBkcm9wLXNlcSBpcyBzdGlsbCBvbmUgb2YgdGhlIG1vc3QgCmNvbW1vbmx5IGFkb3B0ZWQgc2NSTkEtc2VxIHByb3RvY29scywgYXMgZXZpZGVuY2VkIGJ5IHRoZQpsYXJnZSBudW1iZXIgb2YgY2l0YXRpb25zIGZvciBNYWNvc2tvICpldCBhbC4qIAooNC4zMDMgY2l0YXRpb25zIGF0IE5vdmVtYmVyIDMsIDIwMjEpLgoKSW4gdGhpcyBwYXJ0aWN1bGFyIGV4cGVyaW1lbnQsIE1hY29za28gKmV0IGFsLiogc2VxdWVuY2VkIDQ5LDMwMCBjZWxscyBmcm9tIHRoZQptb3VzZSByZXRpbmEsIGlkZW50aWZ5aW5nIDM5IHRyYW5zY3JpcHRpb25hbGx5IGRpc3RpbmN0IGNlbGwgcG9wdWxhdGlvbnMuIFRoZQpleHBlcmltZW50IHdhcyBwZXJmb3JtZWQgaW4gNyBiYXRjaGVzLgoKVGhlIHJhdyBkYXRhIChGQVNUUSBmaWxlcykgb2YgdGhlIGV4cGVyaW1lbnQgY2FuIGJlIHJldHJpZXZlZCBmcm9tIHRoZSAKc2VxdWVuY2luZyByZWFkIGFyY2hpdmUgKFNSQSkgdW5kZXIgdGhlIGFjY2Vzc2lvbiAKW1NSUjE4NTMxNzhdKGh0dHBzOi8vdHJhY2UubmNiaS5ubG0ubmloLmdvdi9UcmFjZXMvc3JhLz9ydW49U1JSMTg1MzE3OCkuIApIb3dldmVyLCBnaXZlbiB0aGUgc2l6ZSBvZiB0aGUgZGF0YSwgd2UgaGF2ZSBwcm92aWRlZCBhIG1vcmUgYW1lbmFibGUgc3Vic2V0Cm9mIHRoZSBkYXRhIG9uIG91ciBsYWIgR2l0SHViIHBhZ2UuCgojIyBEYXRhIHF1YW50aWZpY2F0aW9uIHdpdGggUUMKCldlIHdpbGwgbWFrZSB1c2Ugb2YgClthbGV2aW5dKGh0dHBzOi8vc2FsbW9uLnJlYWR0aGVkb2NzLmlvL2VuL2xhdGVzdC9hbGV2aW4uaHRtbCkgc29mdHdhcmUsIHdoaWNoIAppcyBpbnRlZ3JhdGVkIHdpdGggdGhlIApbc2FsbW9uXShodHRwczovL3NhbG1vbi5yZWFkdGhlZG9jcy5pby9lbi9sYXRlc3Qvc2FsbW9uLmh0bWwpIHF1YW50aWZpY2F0aW9uIApzb2Z0d2FyZSB0byBxdWFudGlmeSBhIHN1YnNldCBvZiByZWFkcyBmb3JtIHRoZSBNYWNvc2tvIGV4cGVyaW1lbnQsIHdoaWxlIApzaW11bGF0YW5lb3VzbHkgcHJvdmlkaW5nIHNvbWUgcXVhbGl0eSBjb250cm9sIGZvciB0aGUgc2VxdWVuY2VkIHJlYWRzLgoKVGhlc2Ugc3RlcHMgYXJlIGltcGxlbWVudGVkIGluIGEgc21hbGwgLnNoZWxsIHNjcmlwdCwgd2hpY2ggY2FuIGJlIG9idGFpbmVkCnVzaW5nIHRoZSBmb2xsb3dpbmcgbGluawpbTWFjb3Nrb19zaF0oLi9sYWIxX3ByZXByb2Nlc3NpbmcvYWxldmluX21hY29za28vcHJlcHJvY2Vzc0Ryb3BzZXFfbGFiLnNoKS4KCiMjIERhdGEgZXhwbG9yYXRpb24gYW5kIHByZXByb2Nlc3NpbmcgaW4gUgoKSGVyZSwgd2UgcHJvdmlkZSBhbiBleGFtcGxlIG9mIGEgc2NyaXB0IGZvciBwZXJmb3JtaW5nIGRhdGEgZXhwbG9yYXRpb24KYW5kIHByZXByb2Nlc3NpbmcgaW4gUiBmb3IgdGhlIE1hY29za28gZGF0YXNldAoKW2xhYjFfTWFjb3Nrb1dvcmtmbG93XSguL2xhYjFfTWFjb3Nrb1dvcmtmbG93Lmh0bWwpCgojIEN1b21vIGRhdGFzZXQKCldlIGhlcmUgbWFrZSB1c2Ugb2YgdGhlIHB1YmxpY2F0aW9uIG9mIEFubmEgQ3VvbW8gZXQgYWwuCihsYXN0IGF1dGhvciBPbGl2ZXIgU3RlZ2xlKSwgd2hpY2ggd2Ugd2lsbCByZWZlciB0byBhcyB0aGUgYEN1b21vIGRhdGFzZXRgLiBUaGUgCnBhcGVyIHRoYXQgZGVzY3JpYmVzIHRoaXMgZGF0YXNldCBjYW4gYmUgZm91bmQgdXNpbmcgdGhpcyAKW2xpbmtdKGh0dHBzOi8vd3d3Lm5hdHVyZS5jb20vYXJ0aWNsZXMvczQxNDY3LTAyMC0xNDQ1Ny16KS4KCkluIHRoZSBleHBlcmltZW50LCB0aGUgYXV0aG9ycyBoYXJ2ZXN0ZWQgaW5kdWNlZCBwbHVyaXBvdGVudCBzdGVtIGNlbGxzIChpUFNDcykKZnJvbSAxMjUgaGVhbHRoeSBodW1hbiBkb25vcnMuIFRoZXNlIGNlbGxzIHdlcmUgdXNlZCB0byBzdHVkeSB0aGUgZW5kb2Rlcm0gCmRpZmZlcmVudGlhdGlvbiBwcm9jZXNzLiBJbiB0aGlzIHByb2Nlc3MsIGlQU0NzIGRpZmZlcmVudGlhdGUgdG8gZW5kb2Rlcm0gY2VsbHMsCmEgcHJvY2VzcyB3aGljaCB0YWtlcyBhcHByb3hpbWF0ZWx5IHRocmVlIGRheXMuIEFzIHN1Y2gsIHRoZSBhdXRob3JzIApjdWx0ZXJlZCB0aGUgaVBTQ3MgY2VsbCBsaW5lcyBhbmQgYWxsb3dlZCBmb3IgZGlmZmVyZW50aWF0aW9uIGZvciB0aHJlZSBkYXlzLiAKRHVyaW5nIHRoZSBleHBlcmltZW50LCBjZWxscyB3ZXJlIGhhcnZlc3RlZCBhdCBmb3VyIGRpZmZlcmVudCB0aW1lIHBvaW50czogCmRheTAgKGRpcmVjdGx5IGF0IHRvIGluY3ViYXRpb24pLCBkYXkxLCBkYXkyIGFuZCBkYXkzLiBLbm93aW5nIHRoZSBwcm9jZXNzIG9mIAplbmRvZGVybSBkaWZmZXJlbnRpYXRpb24sIHRoZXNlIHRpbWUgcG9pbnRzIHNob3VsZCBjb3JyZXNwb25kIHdpdGggZGlmZmVyZW50IApjZWxsIHR5cGVzOiBkYXkwIGFyZSAodW5kaWZmZXJlbnRpYXRlZCkgaVBTQ3MsIGRheTEgYXJlIG1lc2VuZG9kZXJtIGNlbGxzLCBkYXkyCmFyZSAiaW50ZXJtZWRpYXRlIiBjZWxscyBhbmQgZGF5MyBhcmUgZnVsbHkgZGlmZmVyZW50aWF0ZWQgZW5kb2Rlcm0gY2VsbHMuCgpUaGlzIGRhdGFzZXQgd2FzIGdlbmVyYXRlZCB1c2luZyB0aGUgKipTTUFSVC1TZXEyKiogc2NSTkEtc2VxIHByb3RvY29sLgoKVGhlIGZpbmFsIGdvYWwgb2YgdGhlIGV4cGVyaW1lbnQgd2FzIHRvIGNoYXJhY3Rlcml6ZSBwb3B1bGF0aW9uIHZhcmlhdGlvbiBpbiB0aGUKcHJvY2VzcyBvZiBlbmRvZGVybSBkaWZmZXJlbnRpYXRpb24uCgojIyBEYXRhIHF1YW50aWZpY2F0aW9uCgpXZSB3aWxsIG1ha2UgdXNlIG9mIHRoZSAKW3NhbG1vbl0oaHR0cHM6Ly9zYWxtb24ucmVhZHRoZWRvY3MuaW8vZW4vbGF0ZXN0L3NhbG1vbi5odG1sKSBxdWFudGlmaWNhdGlvbiAKc29mdHdhcmUgdG8gcXVhbnRpZnkgYSBzdWJzZXQgb2YgdGhlIGNlbGxzIGZvcm0gdGhlIEN1b21vIGV4cGVyaW1lbnQuIE5vdGUgdGhhdApoZXJlIHdlIGRvIG5vdCBtYWtlIHVzZSBvZiBBbGV2aW4sIGFzIEFsZXZpbiBpcyBvbmx5IHVzZWQgZm9yIGRyb3BsZS1iYXNlZApkYXRhIHN1Y2ggYXMgZHJvcC1zZXEgYW5kIDEwWCBkYXRhLgoKVGhlIHNhbG1vbiBxdWFudGlmaWNhdGlvbiBpcyBpbXBsZW1lbnRlZCBpbiBhIHNtYWxsIC5zaGVsbCBzY3JpcHQsIHdoaWNoIGNhbiAKYmUgb2J0YWluZWQgdXNpbmcgdGhlIGZvbGxvd2luZyBsaW5rCltDdW9tb19zaF0oLi9sYWIxX3ByZXByb2Nlc3Npbmcvc2FsbW9uX2N1b21vL3F1YW50aWZ5X2N1b21vX2xhYi5zaCkuCgojIyBEYXRhIGV4cGxvcmF0aW9uIGFuZCBwcmVwcm9jZXNzaW5nIGluIFIKCkhlcmUsIHdlIHByb3ZpZGUgYSB0ZW1wbGF0ZSBmb3IgYSBzY3JpcHQgZm9yIHBlcmZvcm1pbmcgZGF0YSBleHBsb3JhdGlvbgphbmQgcHJlcHJvY2Vzc2luZyBpbiBSIGZvciB0aGUgQ3VvbW8gZGF0YXNldC4gSXQgaXMgdXAgdG8geW91IHRvIGZpbGwgb3V0CnRoaXMgc2NyaXB0LCBiYXNlZCBvbiB0aGUgY29uY2VwdHMgZGlzY3Vzc2VkIGluIHRoZSB0aGVvcnkgc2Vzc2lvbiBhbmQgdGhlCmZ1Y250aW9ucyBleHBsYWluZWQgZm9yIHRoZSBNYWNvc2tvIGFuYWx5c2lzIQoKW2xhYjFfQ3VvbW9UZW1wbGF0ZV0oLi9sYWIxX0N1b21vVGVtcGxhdGUuaHRtbCkKCkhlcmUsIHdlIHByb3ZpZGUgYSAic29sdXRpb24iIGZpbGUgZm9yIHRoZSBhbmFseXNpcyBvZiB0aGUgQ3VvbW8gZGF0YXNldCBmb3IKdGhlIGZpcnN0IGxhYiBzZXNzaW9uOgoKW2xhYjFfQ3VvbW9Xb3JrZmxvd19odG1sXSguL2xhYjFfcHJlcHJvY2Vzc2luZy9sYWIxX0N1b21vV29ya2Zsb3cuaHRtbCkKW2xhYjFfQ3VvbW9Xb3JrZmxvd19ybWRdKC4vbGFiMV9wcmVwcm9jZXNzaW5nL2xhYjFfQ3VvbW9Xb3JrZmxvdy5SbWQpCgotLS0KCgoKCg==</div>
<div class="footer">
    <hr>
    This work is licensed under the <a href= "https://creativecommons.org/licenses/by-nc-sa/4.0">
    CC BY-NC-SA 4.0</a> licence.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("labSession_1.Rmd");
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
