<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Koen Van den Berge and Jeroen Gilis" />


<title>Data import, quality control and normalization for the Cuomo dataset</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>


<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Single-cell course</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="introSequencingBulk.html">A brief introduction to sequencing technology</a>
    </li>
    <li>
      <a href="singleCellProtocols.html">Two popular single-cell protocols</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Labs
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="labSession_1.html">Lab 1</a>
    </li>
    <li class="dropdown-header">Lab 2</li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/statOmics/Rmd-website">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://statomics.github.io/">statOmics</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
<li role="separator" class="divider"></li>
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Data import, quality control and normalization for the Cuomo dataset</h1>
<h4 class="author">Koen Van den Berge and Jeroen Gilis</h4>
<h4 class="date">24/11/2021</h4>

</div>


<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>We here make use of the publication of Anna Cuomo et al. (last author Oliver Stegle), which we will refer to as the <code>iPSC dataset</code>. The paper that describes this dataset can be found using this <a href="https://www.nature.com/articles/s41467-020-14457-z">link</a>.</p>
<p>In the experiment, the authors harvested induced pluripotent stem cells (iPSCs) from 125 healthy human donors. These cells were used to study the endoderm differentiation process. In this process, iPSCs differentiate to endoderm cells, a process which takes approximately three days. As such, the authors cultered the iPSCs cell lines and allowed for differentiation for three days. During the experiment, cells were harvested at four different time points: day0 (directly at to incubation), day1, day2 and day3. Knowing the process of endoderm differentiation, these time points should correspond with different cell types: day0 are (undifferentiated) iPSCs, day1 are mesendoderm cells, day2 are “intermediate” cells and day3 are fully differentiated endoderm cells.</p>
<p>This dataset was generated using the <strong>SMART-Seq2</strong> scRNA-seq protocol.</p>
<p>The final goal of the experiment was to characterize population variation in the process of endoderm differentiation.</p>
</div>
<div id="download-data" class="section level1">
<h1><span class="header-section-number">2</span> Download data</h1>
<p>For this lab session, we will work with a subset of the data, i.e., the data for the first (alphabetically) 15 patients in the experiment. These can be downloaded through the <em>belnet filesender</em> link provided through email, <a href="https://filesender.belnet.be/?s=download&amp;token=eb8136df-67d3-4869-b2a9-f65767054e81" class="uri">https://filesender.belnet.be/?s=download&amp;token=eb8136df-67d3-4869-b2a9-f65767054e81</a>.</p>
<p>The data original (125 patient) could be downloaded from <a href="https://zenodo.org/record/3625024#.YWfahtlBxB1">Zenodo</a>. At the bottom of this web-page, we can download the files <code>raw_counts.csv.zip</code> and <code>cell_metadata_cols.tsv</code> and store these files locally. We do not recommend doing this during the lab session, to avoid overloading the system.</p>
</div>
<div id="import-data" class="section level1">
<h1><span class="header-section-number">3</span> Import data</h1>
<p>First we read in the count matrix:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>sce &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;/Users/jg/Desktop/sce_15_cuomo.rds&quot;</span>) <span class="co">#change to your data path</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>sce</span></code></pre></div>
</div>
<div id="explore-metadata" class="section level1">
<h1><span class="header-section-number">4</span> Explore metadata</h1>
<p>Exploration of the metadata is essential to get a better idea of what the experiment was about and how it was organized. In contrast with the previous dataset by Macosko et al., we here have a large amount of metadata that we can work with; and that we need to explore.</p>
<p>When we think of the experiment, the key aspects are;</p>
<ul>
<li><p>At which day of the developmental process the cells were sequenced (which should be a proxy for the cell type)</p></li>
<li><p>Cells come from 125 (15 in this reduced dataset) donors</p></li>
</ul>
<p>In addition, to reduce technical artefacts and to allow for batch correction, each batch (“experiment” variable) may contain cells of multiple patients/days.</p>
<p>Explore he metadata. The <code>table()</code> function will come in handy for this (see the Macosko analysis).</p>
</div>
<div id="obtaining-and-including-rowdata" class="section level1">
<h1><span class="header-section-number">5</span> Obtaining and including rowData</h1>
<ul>
<li><p>Assess what is currently stored in the <code>rowData</code> of the SingelCellExperiment object.</p></li>
<li><p>Retrieve relevant information form <code>bioMart</code>. Make sure to select the right values for the <code>dataset</code> and <code>version</code> arguments for the <code>useEnsembl</code> function (these can be retrieved from the Cuomo et al. paper).</p></li>
</ul>
</div>
<div id="filtering-non-informative-genes" class="section level1">
<h1><span class="header-section-number">6</span> Filtering non-informative genes</h1>
<p>Filter the genes using relavant criteria. Compare your results with what we obtained with the Macosko analysis. Can you explain what you observe?</p>
</div>
<div id="quality-control" class="section level1">
<h1><span class="header-section-number">7</span> Quality control</h1>
<div id="calculate-qc-variables" class="section level2">
<h2><span class="header-section-number">7.1</span> Calculate QC variables</h2>
<p>Use <code>perCellQCMetrics</code> to compute QC metrics.</p>
</div>
<div id="exploratory-data-analysis" class="section level2">
<h2><span class="header-section-number">7.2</span> Exploratory data analysis</h2>
</div>
<div id="qc-using-adaptive-thresholds" class="section level2">
<h2><span class="header-section-number">7.3</span> QC using adaptive thresholds</h2>
<p>Visualize the cells that are going to be removed. Are you happy with the selection criterion, i.e., does it appear that we are only removing technical artefacts or could we be removing biological signal as well?</p>
<p>To do this, try coloring the “detected” versus “subsets_Mito_percent” plot and “sum” versus “detected” plots based on biologically significant metadata.</p>
</div>
<div id="remove-empty-droplets" class="section level2">
<h2><span class="header-section-number">7.4</span> Remove empty droplets</h2>
<p>What do you think of this step for the analysis of this dataset?</p>
</div>
<div id="identifying-and-removing-doublets" class="section level2">
<h2><span class="header-section-number">7.5</span> Identifying and removing doublets</h2>
<p>What do you think of this step for the analysis of this dataset?</p>
</div>
</div>
<div id="normalization" class="section level1">
<h1><span class="header-section-number">8</span> Normalization</h1>
</div>

<div id="rmd-source-code">LS0tCnRpdGxlOiAnRGF0YSBpbXBvcnQsIHF1YWxpdHkgY29udHJvbCBhbmQgbm9ybWFsaXphdGlvbiBmb3IgdGhlIEN1b21vIGRhdGFzZXQnCmF1dGhvcjogIktvZW4gVmFuIGRlbiBCZXJnZSBhbmQgSmVyb2VuIEdpbGlzIgpkYXRlOiAiMjQvMTEvMjAyMSIKb3V0cHV0OiAKICBodG1sX2RvY3VtZW50OgogICAgdG9jOiB0cnVlCiAgICB0b2NfZmxvYXQ6IHRydWUKLS0tCgojIEludHJvZHVjdGlvbgoKV2UgaGVyZSBtYWtlIHVzZSBvZiB0aGUgcHVibGljYXRpb24gb2YgQW5uYSBDdW9tbyBldCBhbC4KKGxhc3QgYXV0aG9yIE9saXZlciBTdGVnbGUpLCB3aGljaCB3ZSB3aWxsIHJlZmVyIHRvIGFzIHRoZSBgaVBTQyBkYXRhc2V0YC4gVGhlIApwYXBlciB0aGF0IGRlc2NyaWJlcyB0aGlzIGRhdGFzZXQgY2FuIGJlIGZvdW5kIHVzaW5nIHRoaXMgCltsaW5rXShodHRwczovL3d3dy5uYXR1cmUuY29tL2FydGljbGVzL3M0MTQ2Ny0wMjAtMTQ0NTcteikuCgpJbiB0aGUgZXhwZXJpbWVudCwgdGhlIGF1dGhvcnMgaGFydmVzdGVkIGluZHVjZWQgcGx1cmlwb3RlbnQgc3RlbSBjZWxscyAoaVBTQ3MpCmZyb20gMTI1IGhlYWx0aHkgaHVtYW4gZG9ub3JzLiBUaGVzZSBjZWxscyB3ZXJlIHVzZWQgdG8gc3R1ZHkgdGhlIGVuZG9kZXJtIApkaWZmZXJlbnRpYXRpb24gcHJvY2Vzcy4gSW4gdGhpcyBwcm9jZXNzLCBpUFNDcyBkaWZmZXJlbnRpYXRlIHRvIGVuZG9kZXJtIGNlbGxzLAphIHByb2Nlc3Mgd2hpY2ggdGFrZXMgYXBwcm94aW1hdGVseSB0aHJlZSBkYXlzLiBBcyBzdWNoLCB0aGUgYXV0aG9ycyAKY3VsdGVyZWQgdGhlIGlQU0NzIGNlbGwgbGluZXMgYW5kIGFsbG93ZWQgZm9yIGRpZmZlcmVudGlhdGlvbiBmb3IgdGhyZWUgZGF5cy4gCkR1cmluZyB0aGUgZXhwZXJpbWVudCwgY2VsbHMgd2VyZSBoYXJ2ZXN0ZWQgYXQgZm91ciBkaWZmZXJlbnQgdGltZSBwb2ludHM6IApkYXkwIChkaXJlY3RseSBhdCB0byBpbmN1YmF0aW9uKSwgZGF5MSwgZGF5MiBhbmQgZGF5My4gS25vd2luZyB0aGUgcHJvY2VzcyBvZiAKZW5kb2Rlcm0gZGlmZmVyZW50aWF0aW9uLCB0aGVzZSB0aW1lIHBvaW50cyBzaG91bGQgY29ycmVzcG9uZCB3aXRoIGRpZmZlcmVudCAKY2VsbCB0eXBlczogZGF5MCBhcmUgKHVuZGlmZmVyZW50aWF0ZWQpIGlQU0NzLCBkYXkxIGFyZSBtZXNlbmRvZGVybSBjZWxscywgZGF5MgphcmUgImludGVybWVkaWF0ZSIgY2VsbHMgYW5kIGRheTMgYXJlIGZ1bGx5IGRpZmZlcmVudGlhdGVkIGVuZG9kZXJtIGNlbGxzLgoKVGhpcyBkYXRhc2V0IHdhcyBnZW5lcmF0ZWQgdXNpbmcgdGhlICoqU01BUlQtU2VxMioqIHNjUk5BLXNlcSBwcm90b2NvbC4KClRoZSBmaW5hbCBnb2FsIG9mIHRoZSBleHBlcmltZW50IHdhcyB0byBjaGFyYWN0ZXJpemUgcG9wdWxhdGlvbiB2YXJpYXRpb24gaW4gdGhlCnByb2Nlc3Mgb2YgZW5kb2Rlcm0gZGlmZmVyZW50aWF0aW9uLgoKIyBEb3dubG9hZCBkYXRhCgpGb3IgdGhpcyBsYWIgc2Vzc2lvbiwgd2Ugd2lsbCB3b3JrIHdpdGggYSBzdWJzZXQgb2YgdGhlIGRhdGEsIGkuZS4sIHRoZSBkYXRhCmZvciB0aGUgZmlyc3QgKGFscGhhYmV0aWNhbGx5KSAxNSBwYXRpZW50cyBpbiB0aGUgZXhwZXJpbWVudC4gVGhlc2UgY2FuIGJlCmRvd25sb2FkZWQgdGhyb3VnaCB0aGUgKmJlbG5ldCBmaWxlc2VuZGVyKiBsaW5rIHByb3ZpZGVkIHRocm91Z2ggZW1haWwsCmh0dHBzOi8vZmlsZXNlbmRlci5iZWxuZXQuYmUvP3M9ZG93bmxvYWQmdG9rZW49ZWI4MTM2ZGYtNjdkMy00ODY5LWIyYTktZjY1NzY3MDU0ZTgxLgoKVGhlIGRhdGEgb3JpZ2luYWwgKDEyNSBwYXRpZW50KSBjb3VsZCBiZSBkb3dubG9hZGVkIGZyb20gCltaZW5vZG9dKGh0dHBzOi8vemVub2RvLm9yZy9yZWNvcmQvMzYyNTAyNCMuWVdmYWh0bEJ4QjEpLiBBdCB0aGUgYm90dG9tIG9mIHRoaXMKd2ViLXBhZ2UsIHdlIGNhbiBkb3dubG9hZCB0aGUgZmlsZXMgYHJhd19jb3VudHMuY3N2LnppcGAgYW5kIApgY2VsbF9tZXRhZGF0YV9jb2xzLnRzdmAgYW5kIHN0b3JlIHRoZXNlIGZpbGVzIGxvY2FsbHkuIFdlIGRvIG5vdCByZWNvbW1lbmQgCmRvaW5nIHRoaXMgZHVyaW5nIHRoZSBsYWIgc2Vzc2lvbiwgdG8gYXZvaWQgb3ZlcmxvYWRpbmcgdGhlIHN5c3RlbS4KCiMgSW1wb3J0IGRhdGEKCkZpcnN0IHdlIHJlYWQgaW4gdGhlIGNvdW50IG1hdHJpeDoKCmBgYHtyLCBtZXNzYWdlPUZBTFNFLCB3YXJuaW5nPUZBTFNFLCBldmFsPUZBTFNFfQpzY2UgPC0gcmVhZFJEUygiL1VzZXJzL2pnL0Rlc2t0b3Avc2NlXzE1X2N1b21vLnJkcyIpICNjaGFuZ2UgdG8geW91ciBkYXRhIHBhdGgKc2NlCmBgYAoKIyBFeHBsb3JlIG1ldGFkYXRhCgpFeHBsb3JhdGlvbiBvZiB0aGUgbWV0YWRhdGEgaXMgZXNzZW50aWFsIHRvIGdldCBhIGJldHRlciBpZGVhIG9mIHdoYXQgdGhlCmV4cGVyaW1lbnQgd2FzIGFib3V0IGFuZCBob3cgaXQgd2FzIG9yZ2FuaXplZC4gSW4gY29udHJhc3Qgd2l0aCB0aGUgcHJldmlvdXMKZGF0YXNldCBieSBNYWNvc2tvIGV0IGFsLiwgd2UgaGVyZSBoYXZlIGEgbGFyZ2UgYW1vdW50IG9mIG1ldGFkYXRhIHRoYXQgd2UgY2FuCndvcmsgd2l0aDsgYW5kIHRoYXQgd2UgbmVlZCB0byBleHBsb3JlLgoKV2hlbiB3ZSB0aGluayBvZiB0aGUgZXhwZXJpbWVudCwgdGhlIGtleSBhc3BlY3RzIGFyZTsKCi0gQXQgd2hpY2ggZGF5IG9mIHRoZSBkZXZlbG9wbWVudGFsIHByb2Nlc3MgdGhlIGNlbGxzIHdlcmUgc2VxdWVuY2VkICh3aGljaApzaG91bGQgYmUgYSBwcm94eSBmb3IgdGhlIGNlbGwgdHlwZSkKCi0gQ2VsbHMgY29tZSBmcm9tIDEyNSAoMTUgaW4gdGhpcyByZWR1Y2VkIGRhdGFzZXQpIGRvbm9ycwoKSW4gYWRkaXRpb24sIHRvIHJlZHVjZSB0ZWNobmljYWwgYXJ0ZWZhY3RzIGFuZCB0byBhbGxvdyBmb3IgYmF0Y2ggY29ycmVjdGlvbiwgCmVhY2ggYmF0Y2ggKCJleHBlcmltZW50IiB2YXJpYWJsZSkgbWF5IGNvbnRhaW4gY2VsbHMgb2YgbXVsdGlwbGUgcGF0aWVudHMvZGF5cy4KCkV4cGxvcmUgaGUgbWV0YWRhdGEuIFRoZSBgdGFibGUoKWAgZnVuY3Rpb24gd2lsbCBjb21lIGluIGhhbmR5IGZvciB0aGlzIChzZWUKdGhlIE1hY29za28gYW5hbHlzaXMpLgoKIyBPYnRhaW5pbmcgYW5kIGluY2x1ZGluZyByb3dEYXRhCgotIEFzc2VzcyB3aGF0IGlzIGN1cnJlbnRseSBzdG9yZWQgaW4gdGhlIGByb3dEYXRhYCBvZiB0aGUgU2luZ2VsQ2VsbEV4cGVyaW1lbnQKb2JqZWN0LgoKLSBSZXRyaWV2ZSByZWxldmFudCBpbmZvcm1hdGlvbiBmb3JtIGBiaW9NYXJ0YC4gTWFrZSBzdXJlIHRvIHNlbGVjdCB0aGUgcmlnaHQKdmFsdWVzIGZvciB0aGUgYGRhdGFzZXRgIGFuZCBgdmVyc2lvbmAgYXJndW1lbnRzIGZvciB0aGUgYHVzZUVuc2VtYmxgIGZ1bmN0aW9uCih0aGVzZSBjYW4gYmUgcmV0cmlldmVkIGZyb20gdGhlIEN1b21vIGV0IGFsLiBwYXBlcikuCgojIEZpbHRlcmluZyBub24taW5mb3JtYXRpdmUgZ2VuZXMKCkZpbHRlciB0aGUgZ2VuZXMgdXNpbmcgcmVsYXZhbnQgY3JpdGVyaWEuIENvbXBhcmUgeW91ciByZXN1bHRzIHdpdGggd2hhdCB3ZQpvYnRhaW5lZCB3aXRoIHRoZSBNYWNvc2tvIGFuYWx5c2lzLiBDYW4geW91IGV4cGxhaW4gd2hhdCB5b3Ugb2JzZXJ2ZT8KCiMgUXVhbGl0eSBjb250cm9sCgojIyBDYWxjdWxhdGUgUUMgdmFyaWFibGVzCgpVc2UgYHBlckNlbGxRQ01ldHJpY3NgIHRvIGNvbXB1dGUgUUMgbWV0cmljcy4KCiMjIEV4cGxvcmF0b3J5IGRhdGEgYW5hbHlzaXMKCiMjIFFDIHVzaW5nIGFkYXB0aXZlIHRocmVzaG9sZHMKClZpc3VhbGl6ZSB0aGUgY2VsbHMgdGhhdCBhcmUgZ29pbmcgdG8gYmUgcmVtb3ZlZC4gQXJlIHlvdSBoYXBweSB3aXRoIHRoZQpzZWxlY3Rpb24gY3JpdGVyaW9uLCBpLmUuLCBkb2VzIGl0IGFwcGVhciB0aGF0IHdlIGFyZSBvbmx5IHJlbW92aW5nCnRlY2huaWNhbCBhcnRlZmFjdHMgb3IgY291bGQgd2UgYmUgcmVtb3ZpbmcgYmlvbG9naWNhbCBzaWduYWwgYXMgd2VsbD8KClRvIGRvIHRoaXMsIHRyeSBjb2xvcmluZyB0aGUgImRldGVjdGVkIiB2ZXJzdXMgInN1YnNldHNfTWl0b19wZXJjZW50IiBwbG90IGFuZAoic3VtIiB2ZXJzdXMgImRldGVjdGVkIiBwbG90cyBiYXNlZCBvbiBiaW9sb2dpY2FsbHkgc2lnbmlmaWNhbnQgbWV0YWRhdGEuCgojIyBSZW1vdmUgZW1wdHkgZHJvcGxldHMKCldoYXQgZG8geW91IHRoaW5rIG9mIHRoaXMgc3RlcCBmb3IgdGhlIGFuYWx5c2lzIG9mIHRoaXMgZGF0YXNldD8KCiMjIElkZW50aWZ5aW5nIGFuZCByZW1vdmluZyBkb3VibGV0cwoKV2hhdCBkbyB5b3UgdGhpbmsgb2YgdGhpcyBzdGVwIGZvciB0aGUgYW5hbHlzaXMgb2YgdGhpcyBkYXRhc2V0PwoKIyBOb3JtYWxpemF0aW9uCgoKCgoKCg==</div>
<div class="footer">
    <hr>
    This work is licensed under the <a href= "https://creativecommons.org/licenses/by-nc-sa/4.0">
    CC BY-NC-SA 4.0</a> licence.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("singleCell_CuomoWorkflow_template.Rmd");
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
